{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline for clustering analysis of protein sequences\n",
    "\n",
    "## 1. Dataset\n",
    "- Lysozyme CGCh dataset contains fasta file of Lysozyme C, Lysozyme G, and Lysozyme Ch protein families\n",
    "- Lysozyme C, Lysozyme G, and Lysozyme Ch are labels\n",
    "\n",
    "## 2. Preprocessing\n",
    "- Removes duplicate sequences within and between protein families\n",
    "- Eliminates sequences containing non-standard amino acids (UZOBJX)\n",
    "- Optional: Removes outliers based on sequence length\n",
    "- Optional: Filters sequences below length threshold\n",
    "\n",
    "## 3. Feature Extraction\n",
    "- Uses Pseudo Amino Acid Composition (PAAC)\n",
    "- Considers physicochemical properties:\n",
    "    - Hydrophobicity\n",
    "    - Hydrophilicity\n",
    "    - Mass\n",
    "\n",
    "## 4. Clustering Analysis\n",
    "Implements multiple clustering algorithms:\n",
    "- KMeans\n",
    "- Agglomerative\n",
    "- Gaussian Mixture\n",
    "- Spectral\n",
    "- Affinity Propagation\n",
    "- DBSCAN\n",
    "- OPTICS\n",
    "- Mean-shift\n",
    "- BIRCH\n",
    "\n",
    "## 5. Evaluation Metrics\n",
    "### Intrinsic Metrics\n",
    "- Silhouette Coefficient (SC)\n",
    "- Calinski-Harabasz Index (CH)\n",
    "- Davies-Bouldin Index (DB)\n",
    "\n",
    "### Extrinsic Metrics\n",
    "- Fowlkes-Mallows Index (FMI)\n",
    "- F1 Score\n",
    "- Matthews Correlation Coefficient (MCC)\n",
    "- Adjusted Mutual Information (AMI)\n",
    "- Contingency matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import os\n",
    "os.environ['OMP_NUM_THREADS'] = '1' # set the number of threads to 1 to avoid warning message\n",
    "import time\n",
    "from collections import defaultdict, Counter\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import cluster, mixture, manifold, decomposition, metrics, preprocessing # https://scikit-learn.org/stable/modules/clustering.html\n",
    "import numpy as np\n",
    "import pandas as pd \n",
    "import warnings\n",
    "from sklearn.exceptions import ConvergenceWarning # to suppress warning message \n",
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning, message=\"Number of subclusters found.*by BIRCH.*\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Dataset\n",
    "- Lysozyme CGCh dataset contains fasta file of Lysozyme C, Lysozyme G, and Lysozyme Ch protein families\n",
    "- Lysozyme C, Lysozyme G, and Lysozyme Ch are labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'Lysozyme_CGCCh' # dataset name\n",
    "# protein_families is a dictionary with fasta file path as key and protein family as value\n",
    "protein_families = {'Lysozyme_CGCh/Lysozyme_C.txt':'Lysozyme_C',\n",
    "                    'Lysozyme_CGCh/Lysozyme_G.txt':'Lysozyme_G',\n",
    "                    'Lysozyme_CGCh/Lysozyme_Ch.txt':'Lysozyme_Ch'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1. Preprocessing\n",
    "- Removes duplicate sequences within and between protein families\n",
    "- Eliminates sequences containing non-standard amino acids (UZOBJX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocessing the data\n",
    "class preprocessing_step:\n",
    "    def __init__(self, protein_families):\n",
    "        fasta_sequences = {}\n",
    "        duplicate_accession = {}\n",
    "\n",
    "        # reading the file\n",
    "        for fasta_file, label in protein_families.items():\n",
    "            fasta = []\n",
    "            with open(fasta_file,'r') as file:\n",
    "                for each_line in file:\n",
    "                    fasta.append(each_line.strip())\n",
    "\n",
    "            # convert multiple lines fasta sequence to single line fasta sequence\n",
    "            single_fasta = self.multiple_fasta_to_single_fasta(fasta)\n",
    "\n",
    "            # remove duplicates \"within\" protein family based on accession number\n",
    "            fasta_sequences[label], duplicate_accession[label] = self.remove_intra_family_duplicates(single_fasta,label)\n",
    "            print(f'{label}')\n",
    "            print(f'Total intra-family sequences: {len(single_fasta)//2}, Non-duplicates(accession): {len(fasta_sequences[label])}')\n",
    "\n",
    "        # remove duplicates \"between\" protein families based on accession number\n",
    "        no_duplicate_fasta = self.remove_inter_family_duplicates(fasta_sequences)\n",
    "        # remove sequences with \"UZOBJX\" amino acid\n",
    "        no_uzobjx_fasta = self.remove_UZOBJX(no_duplicate_fasta)\n",
    "        # final preprocessed fasta data\n",
    "        self.all_fasta, self.all_fasta_length = self.final_preprocessed_fasta_data(no_uzobjx_fasta)\n",
    "        print('Total number of protein sequences in the dataset:',len(self.all_fasta)//2)\n",
    "\n",
    "    # function to final get preprocessed fasta data\n",
    "    def final_preprocessed_fasta_data(self, no_uzobjx_fasta):\n",
    "        fasta = []\n",
    "        fasta_length = {}\n",
    "        for i in no_uzobjx_fasta:\n",
    "            fasta_length[i] = []\n",
    "            for j in no_uzobjx_fasta[i]:\n",
    "                fasta.append(j+'$'+i) # store the accession number with labels (protein family) separated by '$'\n",
    "                fasta.append(no_uzobjx_fasta[i][j]) # store the sequence\n",
    "                fasta_length[i].append(len(no_uzobjx_fasta[i][j])) # store the length of the sequence\n",
    "        return fasta, fasta_length\n",
    "\n",
    "    # function to remove sequences with \"UZOBJX\" amino acid\n",
    "    def remove_UZOBJX(self, fasta_sequences):\n",
    "        c = 0\n",
    "        for i in fasta_sequences:\n",
    "            for j in list(fasta_sequences[i]):\n",
    "                if any(aa in fasta_sequences[i][j] for aa in 'UZOBJX'):\n",
    "                    del fasta_sequences[i][j]\n",
    "                    c += 1\n",
    "        print(f'Total sequences after removing UZOBJX: {c}')\n",
    "        return fasta_sequences\n",
    "\n",
    "    # function to remove duplicates between protein families based on accession number\n",
    "    def remove_inter_family_duplicates(self, fasta_sequences):\n",
    "        all_fasta = []\n",
    "        for i in fasta_sequences:\n",
    "            all_fasta += list(fasta_sequences[i])\n",
    "        c = Counter(all_fasta)\n",
    "        duplicates = [i for i in c if c[i]>1] # duplicates between protein families and >1 represents the number of times the duplicates are present\n",
    "        # remove duplicates between protein families\n",
    "        for i in fasta_sequences:\n",
    "            for j in duplicates:\n",
    "                try:\n",
    "                    del fasta_sequences[i][j]\n",
    "                except KeyError:\n",
    "                    pass\n",
    "        all_sequences = {i:len(fasta_sequences[i]) for i in fasta_sequences}\n",
    "        print(f'Total sequences after removing inter-family duplicates:\\n {all_sequences}')\n",
    "        return fasta_sequences       \n",
    "\n",
    "    # function to convert multiple fasta sequence into a single sequence\n",
    "    def multiple_fasta_to_single_fasta(self,multiple_fasta):\n",
    "        single_fasta = []\n",
    "        temp = ''\n",
    "        for i in multiple_fasta:\n",
    "            if i.startswith('>'):\n",
    "                single_fasta.append(temp)\n",
    "                single_fasta.append(i.rstrip())\n",
    "                temp = ''\n",
    "            else:\n",
    "                temp += i.rstrip()\n",
    "        single_fasta.append(temp)\n",
    "        return single_fasta[1:]\n",
    "    \n",
    "    # function to remove duplicates within protein family based on accession number\n",
    "    def remove_intra_family_duplicates(self, single_fasta, label):\n",
    "        protein_family = {}\n",
    "        duplicates = []\n",
    "        # converted the fasta sequence into a dictionary with key as accession number and value as the sequence\n",
    "        for i in range(0,len(single_fasta),2):\n",
    "            try:\n",
    "                protein_family[single_fasta[i]] # check if the accession number exists, if yes then it is a duplicate\n",
    "                duplicates.append(single_fasta[i]) # store the accession number duplicates\n",
    "            except KeyError:\n",
    "                protein_family[single_fasta[i]] = single_fasta[i+1].upper()\n",
    "        return protein_family, duplicates\n",
    "\n",
    "# preprocessed_fasta = preprocessing_step(protein_families) # protein_families is a dictionary with fasta file as key and protein family as value\n",
    "# preprocessed_fasta_data.all_fasta # Contains the preprocessed fasta data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. Preprocessing\n",
    "- Optional: Removes outliers based on sequence length\n",
    "- Optional: Filters sequences below length threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional preprocessing step: \n",
    "# 1) Remove outliers based on the lower and upper bounds for the protein sequences length using box-whisker plot\n",
    "# 2) Remove the sequence below assigned sequence length cut-off\n",
    "class optional_preprocessing_step:\n",
    "    def __init__(self, data, len_cut_off):\n",
    "        # get the outliers, lower and upper bounds for the protein sequences length using box-whisker plot\n",
    "        box_cut_off = self.box_whisker_plot(data.all_fasta_length)\n",
    "        print('Outlier cut-off for each protein family:', box_cut_off)\n",
    "        # remove outliers (sequences) based on the lower and upper bounds from fasta data\n",
    "        self.no_outlier_fasta = self.remove_outliers(data.all_fasta, box_cut_off)\n",
    "        #remove the sequence below assigned cut-off\n",
    "        self.len_filtered_fasta = self.remove_below_cut_off(self.no_outlier_fasta, len_cut_off)\n",
    "\n",
    "    # function to remove sequences below the cut-off  \n",
    "    def remove_below_cut_off(self, fasta_data, len_cut_off):\n",
    "        fasta = []\n",
    "        for i in range(1,len(fasta_data),2):\n",
    "            if len(fasta_data[i]) >= len_cut_off:\n",
    "                pass\n",
    "                fasta.append(fasta_data[i-1])\n",
    "                fasta.append(fasta_data[i])\n",
    "        print(f'Total number of protein sequences below cut-off {len_cut_off}: {len(fasta_data)//2 - len(fasta)//2}')\n",
    "        return fasta\n",
    "\n",
    "    # function to remove outliers based on the lower and upper bounds from fasta data\n",
    "    def remove_outliers(self, fasta_data, box_cut_off):\n",
    "        fasta = []\n",
    "        for i in range(0,len(fasta_data),2):\n",
    "            j = fasta_data[i].split('$')[-1] # get the protein family\n",
    "            if len(fasta_data[i+1]) >= box_cut_off[j][0] and len(fasta_data[i+1]) <= box_cut_off[j][1]:\n",
    "                fasta.append(fasta_data[i])\n",
    "                fasta.append(fasta_data[i+1])\n",
    "        print('Total number of protein sequences removed as outliers:',len(fasta_data)//2 - len(fasta)//2)\n",
    "        return fasta\n",
    "        \n",
    "    # function to plot box-whisker plot for protein sequence length and get the lower and upper bounds for each protein family\n",
    "    def box_whisker_plot(self, fasta_length):\n",
    "        fig =  plt.figure(figsize = (6, 3))\n",
    "        box_plot_data = plt.boxplot([fasta_length[i] for i in protein_families.values()], tick_labels = protein_families.values())\n",
    "        plt.xlabel('Protein families')\n",
    "        plt.ylabel('Sequence length')\n",
    "        plt.title('Box-whisker plot for protein sequence length')\n",
    "        plt.show()\n",
    "   \n",
    "        cut_off = {}\n",
    "        # get the lower and upper bounds for each protein family based on sequence length\n",
    "        caps = [item.get_ydata() for item in box_plot_data['caps']] # get the lower and upper bounds\n",
    "        cap_index = 0\n",
    "        for family in protein_families.values():\n",
    "            bounds = []\n",
    "            for cap in caps[cap_index:cap_index+2]:\n",
    "                bounds.append(cap[0])\n",
    "            cut_off[family] = bounds\n",
    "            cap_index += 2\n",
    "        return cut_off\n",
    "        \n",
    "# optional_preprocessing_fasta = optional_preprocessing_step(preprocessed_fasta, 200) # 200 is the cut-off for the protein sequence length   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Feature Extraction\n",
    "- Uses Pseudo Amino Acid Composition (PAAC)\n",
    "- Considers physicochemical properties:\n",
    "    - Hydrophobicity\n",
    "    - Hydrophilicity\n",
    "    - Mass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Feature extraction using pseudo amino acid composition (PAAC)\n",
    "\n",
    "# Physicochemical properties of amino acids like hydrophobicity, hydrophilicity and mass\n",
    "physicochemical_properties = {}\n",
    "physicochemical_properties.update({\n",
    "    'hydrophobicity': {'A':0.62,'C':0.29,'D':-0.09,'E':-0.74,'F':1.19,'G':0.48,'H':-0.4,'I':1.38,'K':-1.5,'L':1.53,'M':0.64,'N':-0.78,'P':0.12,'Q':-0.85,'R':-2.53,'S':-0.18,'T':-0.05,'V':1.8,'W':0.81,'Y':0.26},\n",
    "    'hydrophilicity': {'A':-0.5,'C':-1.0,'D':3.0,'E':3.0,'F':-2.5,'G':0.0,'H':-0.5,'I':-1.8,'K':3.0,'L':-1.8,'M':-1.3,'N':0.2,'P':0.0,'Q':0.2,'R':3.0,'S':0.3,'T':-0.4,'V':-1.5,'W':-3.4,'Y':-2.3},\n",
    "    'mass': {'A':15.0,'C':47.0,'D':59.0,'E':73.0,'F':91.0,'G':1.0,'H':82.0,'I':57.0,'K':73.0,'L':57.0,'M':75.0,'N':58.0,'P':42.0,'Q':72.0,'R':101.0,'S':31.0,'T':45.0,'V':43.0,'W':130.0,'Y':107.0}\n",
    "}) # hydrophobicity, hydrophilicity and mass of amino acids respectively taken from http://www.csbio.sjtu.edu.cn/bioinf/PseAAC/ParaValue.htm\n",
    "\n",
    "# function to standardize the physicochemical properties\n",
    "def standardize_(physicochemical_properties):\n",
    "    new_physicochemical_properties = {}\n",
    "    for key, value in physicochemical_properties.items():\n",
    "        amino_acids = list(value.keys())\n",
    "        original_values = list(value.values())\n",
    "        standardized_values = preprocessing.StandardScaler().fit_transform(np.array(original_values).reshape(-1,1)).flatten()\n",
    "        new_physicochemical_properties[key] = dict(zip(amino_acids, standardized_values))\n",
    "    return new_physicochemical_properties\n",
    "\n",
    "standardized_physicochemical_properties = standardize_(physicochemical_properties)\n",
    "\n",
    "        \n",
    "# Pseudo amino acid composition (PAAC) feature extraction\n",
    "class pseudo_amino_acid_composition:\n",
    "    def __init__(self, fasta_data,lambda_, w):\n",
    "        self.lambda_ = lambda_\n",
    "        self.w = w\n",
    "        self.physicochemical_properties = standardized_physicochemical_properties\n",
    "        self.amino_acids = {1:'A', 2:'C', 3:'D', 4:'E', 5:'F', 6:'G', 7:'H', 8:'I', 9:'K', 10:'L', 11:'M', 12:'N', 13:'P', 14:'Q', 15:'R', 16:'S', 17:'T', 18:'V', 19:'W', 20:'Y'}\n",
    "        collect_features = []\n",
    "        collect_features.append(['#'] + [aa for aa in self.amino_acids.values()] + [f'\\u03BB{i}' for i in range(1,self.lambda_+1)]) # header of the dataframe\n",
    "        for i in range(0,len(fasta_data),2):\n",
    "            accessions = fasta_data[i].rstrip()\n",
    "            paac = self.calculate_paac(fasta_data[i+1].rstrip())\n",
    "            collect_features.append([accessions]+paac)\n",
    "        self.df = pd.DataFrame(collect_features[1:], columns=collect_features[0])\n",
    "        # display(self.df)\n",
    "\n",
    "    # function to calculate pseudo amino acid composition (PAAC)\n",
    "    def calculate_paac(self, sequence):\n",
    "        # Calculating set of sequence order-correlation factors (lower_theta)\n",
    "        lower_theta = self.calculate_lower_theta(sequence)\n",
    "        # calculating denominator\n",
    "        denominator = 1+(self.w*sum(lower_theta.values()))\n",
    "        # calculating pseudo amino acid composition (PAAC)\n",
    "        paac = []\n",
    "        for i in range(1,21+self.lambda_):\n",
    "            if i>=1 and i <= 20: # for amino acids composition of 20 amino acids\n",
    "                numerator = sequence.count(self.amino_acids[i])/len(sequence)\n",
    "                paac.append(numerator/denominator)\n",
    "            elif i >= 21 and i <= 20+self.lambda_: # for sequence order-correlation factors\n",
    "                numerator = self.w*lower_theta[i-20] \n",
    "                paac.append(numerator/denominator) \n",
    "        return paac      \n",
    "\n",
    "    # function to calculate the sequence order-correlation factors (lower_theta)\n",
    "    def calculate_lower_theta(self, sequence):\n",
    "        lower_theta = {}\n",
    "        for i in range(1,self.lambda_+1):\n",
    "            # calculate the correlation function (upper_theta) using physicochemical properties\n",
    "            lower_theta[i] = (1/(len(sequence)-i))*self.calculate_upper_theta(sequence,i)\n",
    "        return lower_theta\n",
    "    \n",
    "    # function to calculate the correlation function (upper_theta) using physicochemical properties\n",
    "    def calculate_upper_theta(self, sequence, i):\n",
    "        upper_theta = []\n",
    "        for j in range(len(sequence)-i):\n",
    "            physicochemical_diff = []\n",
    "            for k in self.physicochemical_properties:\n",
    "                # calculate the difference between the physicochemical properties of amino acids\n",
    "                l = ((self.physicochemical_properties[k][sequence[j]] - self.physicochemical_properties[k][sequence[j+i]])**2)\n",
    "                physicochemical_diff.append(l)\n",
    "            upper_theta.append(sum(physicochemical_diff)/len(self.physicochemical_properties))\n",
    "        return sum(upper_theta)\n",
    "                \n",
    "# PAAC = pseudo_amino_acid_composition(optional_preprocessing_fasta.len_filtered_fasta, 30, 0.05) #"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Clustering Analysis\n",
    "Implements multiple clustering algorithms:\n",
    "- KMeans, Agglomerative, Gaussian Mixture, Spectral\n",
    "- Affinity Propagation, DBSCAN, OPTICS, Mean-shift, BIRCH\n",
    "\n",
    "## 5. Evaluation Metrics\n",
    "### Intrinsic Metrics\n",
    "- Silhouette Coefficient (SC), Calinski-Harabasz Index (CH), Davies-Bouldin Index (DB)\n",
    "\n",
    "### Extrinsic Metrics\n",
    "- Fowlkes-Mallows Index (FMI), F1 Score, Matthews Correlation Coefficient (MCC), Adjusted Mutual Information (AMI)\n",
    "- Contingency matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustering \n",
    "class clustering:\n",
    "    def __init__(self, df, k):\n",
    "        # random seed\n",
    "        self.seed = 77\n",
    "        # separate the accession number and the PAAC features\n",
    "        self.x_values = df.iloc[:,1:].values\n",
    "        self.y_values = df.iloc[:,0].values\n",
    "        self.k = k # number of clusters\n",
    "        self.labels = [i.split('$')[-1] for i in self.y_values] # labels were added to the accession number during preprocessing step (in final_preprocessed_fasta_data function)\n",
    "        sorted_labels = sorted(set(self.labels)) # sort the labels\n",
    "        self.true_labels = [sorted_labels.index(i) for i in self.labels] # true labels with 1,2,3,....n labels\n",
    "\n",
    "    # function of different clustering methods\n",
    "    # 1) KMeans\n",
    "    def kmeans(self):\n",
    "        self.name = 'KMeans'\n",
    "        km = cluster.KMeans(n_clusters=self.k, random_state=self.seed).fit(self.x_values)\n",
    "        self.predicted_labels = km.predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 2) Agglomerative clustering\n",
    "    def agglomerative(self):\n",
    "        self.name = 'Agglomerative'\n",
    "        ag = cluster.AgglomerativeClustering(n_clusters=self.k)\n",
    "        self.predicted_labels = ag.fit_predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 3) Gaussian mixture\n",
    "    def gaussian_mixture(self):\n",
    "        self.name = 'Gaussian mixture'\n",
    "        gmm = mixture.GaussianMixture(n_components=self.k, random_state=self.seed).fit(self.x_values)\n",
    "        self.predicted_labels = gmm.predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 4) Spectral clustering\n",
    "    def spectral(self):\n",
    "        self.name = 'Spectral'\n",
    "        sc = cluster.SpectralClustering(n_clusters=self.k, affinity='nearest_neighbors', assign_labels='discretize' ,random_state=self.seed).fit(self.x_values)\n",
    "        self.predicted_labels = sc.labels_\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 5) Affinity propagation\n",
    "    def affinity_propagation(self):\n",
    "        self.name = 'Affinity propagation'\n",
    "        negative_sq_dist = -metrics.pairwise_distances(self.x_values, metric='sqeuclidean') # calculate the negative square of the euclidean distance\n",
    "        af = cluster.AffinityPropagation(preference=np.min(negative_sq_dist), random_state=self.seed).fit(self.x_values) # preference is the minimum of the square of the euclidean distance (similarity matrix)\n",
    "        self.predicted_labels = af.predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 6) DBSCAN\n",
    "    def dbscan(self):\n",
    "        self.name = 'DBSCAN'\n",
    "        standardize_x_values = preprocessing.StandardScaler().fit_transform(self.x_values) # standardize the x values\n",
    "        db = cluster.DBSCAN()\n",
    "        self.predicted_labels = db.fit_predict(standardize_x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 7) OPTICS\n",
    "    def optics(self):\n",
    "        self.name = 'OPTICS'\n",
    "        op = cluster.OPTICS()\n",
    "        self.predicted_labels = op.fit_predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "\n",
    "    # 8) Mean-shift\n",
    "    def mean_shift(self):\n",
    "        self.name = 'Mean-shift'\n",
    "        ms = cluster.MeanShift().fit(self.x_values)\n",
    "        self.predicted_labels = ms.predict(self.x_values)\n",
    "        return self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "    \n",
    "    # 9) BIRCH\n",
    "    def birch(self):\n",
    "        self.name = 'BIRCH'\n",
    "        br = cluster.Birch(n_clusters=self.k).fit(self.x_values)\n",
    "        self.predicted_labels = br.predict(self.x_values)\n",
    "        self. metrics = self.metrics() # return dictionary of info about the clustering method and evaluation metrics\n",
    "        print(f'Number of subclusters found {len(set(self.predicted_labels))} by BIRCH even though k={self.k}')\n",
    "        return self.metrics\n",
    "        \n",
    "    #  function to calculate evaluation metrics\n",
    "    def metrics(self):\n",
    "        # Intrinsic evaluation metrics\n",
    "        predicted_k = len(set(self.predicted_labels)) - (1 if -1 in self.predicted_labels else 0) # -1 represents the noise points in case of DBSCAN, OPTICS\n",
    "\n",
    "        SC = metrics.silhouette_score(self.x_values, self.predicted_labels, metric='euclidean') if predicted_k > 1 else None # Silhouette coefficient\n",
    "        CH = metrics.calinski_harabasz_score(self.x_values, self.predicted_labels) if predicted_k > 1 else None  # Calinski-Harabasz index\n",
    "        DB = metrics.davies_bouldin_score(self.x_values, self.predicted_labels) if predicted_k > 1 else None # Davies-Bouldin index\n",
    "        \n",
    "        # Extrinsic evaluation metrics\n",
    "        # Compare pairs of labels within the true and predicted labels\n",
    "        true_label_pairs = self.get_label_pairs(self.true_labels) # true label pairs, 1 indicates same cluster, 0 indicates different cluster   \n",
    "        predicted_label_pairs = self.get_label_pairs(self.predicted_labels) # predicted label pairs, 1 indicates same cluster, 0 indicates different cluster\n",
    "        # Compare pairs of labels between the true and predicted labels\n",
    "        TP = np.dot(true_label_pairs, predicted_label_pairs).astype(float) # True positive\n",
    "        TN = np.dot((true_label_pairs == 0).astype(float), (predicted_label_pairs == 0).astype(float)) # True negative\n",
    "        FP = np.dot((true_label_pairs == 0).astype(float), predicted_label_pairs) # False positive\n",
    "        FN = np.dot(true_label_pairs, (predicted_label_pairs == 0).astype(float))\n",
    "        # Normalize confusion matrix values\n",
    "        total = TP + TN + FP + FN\n",
    "        TP, TN, FP, FN = TP/total, TN/total, FP/total, FN/total\n",
    "        # Calculate evaluation metrics\n",
    "        FMI = metrics.fowlkes_mallows_score(self.true_labels, self.predicted_labels) # Fowlkes-Mallows index\n",
    "        accuracy = (TP + TN) / (TP + TN + FP + FN) # Accuracy\n",
    "        precision = TP/(TP+FP) # Precision\n",
    "        recall = TP/(TP+FN) # Recall\n",
    "        F1 = 2 * precision * recall / (precision + recall) if (precision + recall) != 0 else 0  # F1 score\n",
    "        MCC = ((TP * TN) - (FP * FN)) / np.sqrt((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN)) if (TP + FP) * (TP + FN) * (TN + FP) * (TN + FN) != 0 else 0  # Matthews correlation coefficient\n",
    "        AMI = metrics.adjusted_mutual_info_score(self.true_labels, self.predicted_labels)\n",
    "        # SC = Silhouette coefficient, CH = Calinski-Harabasz index, DB = Davies-Bouldin index, FMI = Fowlkes-Mallows index, F1 = F1 score, MCC = Matthews correlation coefficient, AMI = Adjusted mutual information\n",
    "        metrics_dict = {'SC': SC, 'CH': CH, 'DB': DB, 'FMI': FMI, 'F1': F1, 'MCC': MCC, 'AMI': AMI}\n",
    "        metrics_dict = {k: round(v, 2) if v is not None else None for k, v in metrics_dict.items()}\n",
    "        \n",
    "        # contingency matrix\n",
    "        contingency_matrix = metrics.cluster.contingency_matrix(self.true_labels, self.predicted_labels) \n",
    "        # Create DataFrame for the contingency matrix with rows as protein families and columns as clusters\n",
    "        df_contingency = pd.DataFrame(contingency_matrix, \n",
    "                                      index=[f'{i}' for i in sorted(set(self.labels))])\n",
    "        # Add noise column if -1 is present in the predicted labels\n",
    "        df_contingency.columns = [f'Cluster {i}' for i in range(predicted_k)] if -1 not in self.predicted_labels else ['Noise']+[f'Cluster {i}' for i in range(predicted_k)]\n",
    "        df_contingency.index.name = 'Protein family'\n",
    "        df_contingency.columns.name = 'Cluster'\n",
    "        df_contingency.loc['Total'] = df_contingency.sum() # Add total row\n",
    "        df_contingency['Total'] = df_contingency.sum(axis=1) # Add total column\n",
    "        print(f'\\nContingency Matrix for {self.name}:')\n",
    "        display(df_contingency)\n",
    "        # info dictionary\n",
    "        info_dict = {'Clustering_method': self.name, 'expected_clusters': len(set(self.true_labels)), 'predicted_clusters': len(set(self.predicted_labels)), \n",
    "                     'total_features': self.x_values.shape[1], 'total_samples': self.x_values.shape[0]}\n",
    "        info_dict.update(metrics_dict) # add evaluation metrics to the dictionary\n",
    "        return info_dict\n",
    "        \n",
    "    # function to get label pairs\n",
    "    def get_label_pairs(self, labels):\n",
    "        # Create binary matrix where 1 indicates same cluster, 0 indicates different cluster\n",
    "        label_similarity_matrix = ((np.expand_dims(labels, 0) == np.expand_dims(labels, 1)).astype(int))\n",
    "        # Get upper triangular pairs (excluding diagonal)\n",
    "        label_pairs = label_similarity_matrix[np.triu_indices(len(labels), k=1)]\n",
    "        return label_pairs\n",
    "\n",
    "# clustering_results = clustering(PAAC.df, 3).kmeans() # 3 is the number of clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output (displayed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clustering results for Lysozyme_CGCCh dataset\n",
      "Lysozyme_C\n",
      "Total intra-family sequences: 64, Non-duplicates(accession): 64\n",
      "Lysozyme_G\n",
      "Total intra-family sequences: 21, Non-duplicates(accession): 21\n",
      "Lysozyme_Ch\n",
      "Total intra-family sequences: 58, Non-duplicates(accession): 58\n",
      "Total sequences after removing inter-family duplicates:\n",
      " {'Lysozyme_C': 64, 'Lysozyme_G': 21, 'Lysozyme_Ch': 58}\n",
      "Total sequences after removing UZOBJX: 0\n",
      "Total number of protein sequences in the dataset: 143\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhwAAAE6CAYAAABd44vKAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/TGe4hAAAACXBIWXMAAA9hAAAPYQGoP6dpAABSKElEQVR4nO3deXxMV/8H8M9kn6xkX6RZBEESaq0QJLGvEUGrllS1HmsRbYVai5Ra2p8+RZ8+kqqWagSV2mJLE0KJKlEkVCyViAZZiKzn94cnt0YWMzEj2+f9es2Le+65937vnTsz39x7zrkyIYQAERERkQZpVXcAREREVPcx4SAiIiKNY8JBREREGseEg4iIiDSOCQcRERFpHBMOIiIi0jgmHERERKRxTDiIiIhI45hwEBERkcYx4ajlIiIiIJPJFF5WVlbo3r07oqOjqzs8lTg7O2PAgAHPrXf06FHIZDIcPXpUI+uvbsHBwXB2dq7Ssl9++SUiIiKUrn/v3j28/vrrsLa2hkwmQ0BAQJW2WxPt2bMHCxcufKF1vMh7QS9P9+7d0b179+oOQ1LR57D0uysyMvLlB1UDMOGoI8LDw5GQkIDjx4/jq6++gra2NgYOHIjdu3dXd2hq16ZNGyQkJKBNmzbVHUqNo2rC8fHHH2PHjh1Ys2YNEhISsGLFCs0F95Lt2bMHixYteqF1zJs3Dzt27FBTRFRfqPo5rC90qjsAUg8PDw+0a9dOmu7Tpw8aNmyILVu2YODAgdUYmfqZmpritddeq+4wynj06BEMDQ2rOwyVJCUloXHjxnjzzTfVsj4hBB4/fgy5XK6W9T2tsLAQMpkMOjov72urcePGL21bRHUdr3DUUQYGBtDT04Ourq5C+b179zBp0iQ4ODhAT08Prq6umDt3LvLz8wEAjx8/xquvvgo3NzdkZWVJy6Wnp8PW1hbdu3dHcXFxhdu9cOECZDIZfvzxR6ksMTERMpkMLVu2VKg7aNAgtG3btsw69u3bhzZt2kAul8Pd3R0bN25UmF/eLZU///wTr7/+Ouzt7aGvrw8bGxv4+/vj7NmzlR6nL7/8Ejo6OliwYIFUdvDgQfj7+8PU1BSGhobo3LkzDh06pLDcwoULIZPJcObMGQQFBaFhw4aV/jiV3vqKiYnBW2+9BXNzcxgZGWHgwIH4888/K40RePK+hIaGwsXFBXp6enBwcMDkyZPx4MEDqY6zszMuXLiA2NhY6fZaRbcDUlNTIZPJcPDgQVy8eFGqX3pMn3eelJLJZJgyZQrWr1+P5s2bQ19fH998802F+1F6W2vHjh3w8vKCgYEBXF1d8X//938K9Urf42+//RYhISFwcHCAvr4+rly5AgDYuHEjWrVqBQMDA5ibm2PIkCG4ePGitHxwcDD+/e9/SzGWvlJTUwE8SYy+/PJLtG7dGnK5HA0bNkRQUFCZ96K8Wyql+/ztt9+iefPmMDQ0RKtWrZS6hVlSUoIlS5agWbNmkMvlaNCgAby8vPD5558r1EtJScHIkSNhbW0NfX19NG/eXNqfp126dAl9+vSBoaEhLC0t8a9//Qu7d+8u8/lwdnZGcHBwmeXLuxWRnZ2NWbNmKZxr06dPx8OHD6t8HC5duoQ33ngDNjY20NfXxyuvvIIxY8YonE/p6emYMGECGjVqBD09Pbi4uGDRokUoKip67nEtT0FBAZYsWQJ3d3fo6+vDysoKb731Fu7evatQr/ScfN73DgDEx8ejU6dOMDAwgIODA+bNm4evv/5a4dxS5nNYWFiIuXPnwt7eHqampujRowcuX75cpf2sVQTVauHh4QKAOHHihCgsLBQFBQXi5s2bYtq0aUJLS0vs27dPqpuXlye8vLyEkZGRWLlypThw4ICYN2+e0NHREf369ZPqJScnCxMTExEYGCiEEKK4uFj4+fkJa2trcfv27efGZGdnJ959911p+pNPPhFyuVwAEH/99ZcQQojCwkJhamoqPvjgA6mek5OTaNSokWjRooXYtGmT2L9/vxg2bJgAIGJjY6V6R44cEQDEkSNHpLJmzZoJNzc38e2334rY2Fixfft2ERISolDHyclJ9O/fXwghRElJiQgJCRG6uroiPDxcqvPtt98KmUwmAgICRFRUlNi9e7cYMGCA0NbWFgcPHpTqLViwQAAQTk5O4sMPPxQxMTFi586dFR6T0vfJ0dFRjBs3Tuzdu1d89dVXwtraWjg6Oor79+9LdceOHSucnJyk6ZKSEtG7d2+ho6Mj5s2bJw4cOCBWrlwpjIyMxKuvvioeP34shBDizJkzwtXVVbz66qsiISFBJCQkiDNnzpQbz+PHj0VCQoJ49dVXhaurq1Q/KytL6fNECCEACAcHB+Hl5SW+//57cfjwYZGUlFThcXBychIODg7ilVdeERs3bhR79uwRb775pgAgPv30U6le6Xvs4OAggoKCxE8//SSio6NFZmamWLZsmQAg3njjDfHzzz+LTZs2CVdXV2FmZiaSk5OFEEJcuXJFBAUFCQDSviUkJEjH6p133hG6uroiJCRE7Nu3T3z//ffC3d1d2NjYiPT09Arfi9J9dnZ2Fh06dBDbtm0Te/bsEd27dxc6Ojri6tWrFe67EEKEhYUJbW1tsWDBAnHo0CGxb98+8dlnn4mFCxdKdS5cuCDMzMyEp6en2LRpkzhw4IAICQkRWlpaCvXS09OFtbW1cHBwEOHh4dKxfOWVV8p8PpycnMTYsWPLxNOtWzfRrVs3afrhw4eidevWwtLSUqxevVocPHhQfP7558LMzEz4+fmJkpISlY/D2bNnhbGxsXB2dhbr168Xhw4dEps3bxbDhw8X2dnZQggh0tLShKOjo3BychIbNmwQBw8eFB9//LHQ19cXwcHBlR7T8vajuLhY9OnTRxgZGYlFixaJmJgY8fXXXwsHBwfRokUL8ejRI4Vjo8z3zu+//y4MDAyEl5eX2Lp1q/jpp59Ev379hLOzswAgrl27JoSo/HNYel47OzuLN998U/z8889iy5Yt4pVXXhFNmjQRRUVFz93X2owJRy1X+kP27EtfX198+eWXCnXXr18vAIht27YplC9fvlwAEAcOHJDKfvjhBwFAfPbZZ2L+/PlCS0tLYX5lRo0aJVxdXaXpHj16iHfeeUc0bNhQfPPNN0IIIY4dO1Zmm05OTsLAwEBcv35dKsvLyxPm5uZiwoQJUtmzCcfff/8txVqZ0oTj0aNHYujQocLMzEwhiXj48KEwNzcXAwcOVFiuuLhYtGrVSnTo0EEqK0045s+fr9QxKX2fhgwZolBeehyWLFkilT37I7dv3z4BQKxYsUJh2dL36KuvvpLKWrZsqfDF+zzdunUTLVu2VChT5TwBIMzMzMS9e/eU2p6Tk5OQyWTi7NmzCuU9e/YUpqam4uHDh0KIf97jrl27KtS7f/++kMvlZRKfGzduCH19fTFy5EipbPLkyaK8v6kSEhIEALFq1SqF8ps3bwq5XK6QBFeUcNjY2Eg/lkI8+fHX0tISYWFhle7/gAEDROvWrSut07t3b9GoUSORlZWlUD5lyhRhYGAgHesPP/ywwmNZ1YQjLCxMaGlpiVOnTinUi4yMFADEnj17pDJlj4Ofn59o0KCByMjIqHCfJ0yYIIyNjRU++0IIsXLlSgFAXLhwocJly9uPLVu2CABi+/btCvVOnTolACh8Nyr7vTNs2DBhZGQk7t69K5UVFxeLFi1aKCQcQlT8OSw9r589f7dt2yYlx3UZb6nUEZs2bcKpU6dw6tQp7N27F2PHjsXkyZPxxRdfSHUOHz4MIyMjBAUFKSxbeqn16dsGw4cPx8SJE/H+++9jyZIlmDNnDnr27CnNF0KgqKhI4VXK398ff/75J65du4bHjx8jPj4effr0ga+vL2JiYgA8uW2hr6+PLl26KMTSunVrvPLKK9K0gYEBmjZtiuvXr1e47+bm5mjcuDE+/fRTrF69Gr/99htKSkrKrZuZmQk/Pz/8+uuviI+Ph7+/vzTv+PHjuHfvHsaOHauwXyUlJejTpw9OnTpV5rLy0KFDK4yrPM+2lfD29oaTkxOOHDlS4TKHDx8GgDKXxIcNGwYjI6Myt3telCrnCQD4+fmhYcOGSq+/ZcuWaNWqlULZyJEjkZ2djTNnziiUP3t8ExISkJeXV+ZYODo6ws/PT6ljER0dDZlMhlGjRim8z7a2tmjVqpVSvZ98fX1hYmIiTdvY2MDa2rrS8xQAOnTogN9//x2TJk3C/v37kZ2drTD/8ePHOHToEIYMGQJDQ0OF+Pr164fHjx/jxIkTAIAjR45UeCyrKjo6Gh4eHmjdurXCtnv37l1uz7DnHYdHjx4hNjYWw4cPh5WVVaXb9fX1hb29vcJ2+/btCwCIjY1VeT8aNGiAgQMHKqyvdevWsLW1LbMfynzvxMbGws/PD5aWllKZlpYWhg8frlJswJPbyU/z8vICgOeeP7UdE446onnz5mjXrh3atWuHPn36YMOGDejVqxc++OAD6T5/ZmYmbG1tIZPJFJa1traGjo4OMjMzFcrHjRuHwsJC6OjoYNq0aQrzYmNjoaurq/AqvYfZo0cPAE+Sivj4eBQWFsLPzw89evSQfhAOHjyIzp07l2lcaGFhUWbf9PX1kZeXV+G+y2QyHDp0CL1798aKFSvQpk0bWFlZYdq0acjJyVGom5ycjJMnT6Jv377w8PBQmHfnzh0AQFBQUJl9W758OYQQuHfvnsIydnZ2FcZVHltb23LLnj32T8vMzISOjk6ZL2yZTPbcZatC1fNEXcegdNuVrbt0fnnbtLe3V+pY3LlzB0II2NjYlHmfT5w4gb///vu566jKeQoAoaGhWLlyJU6cOIG+ffvCwsIC/v7+OH36tLR/RUVFWLt2bZnY+vXrBwBSfKXv07PKK1PWnTt3cO7cuTLbNjExgRCizLF53nG4f/8+iouL0ahRo+dud/fu3WW2W9ruS5n35Nn1PXjwQGrH9vQrPT1d5f0AnhxvGxubMvXKK3ueZ7enr68PAM89f2o79lKpw7y8vLB//34kJyejQ4cOsLCwwMmTJyGEUPgxycjIQFFRkULm/vDhQ4wePRpNmzbFnTt3MH78eOzatUua37ZtW5w6dUphe/b29gCARo0aoWnTpjh48CCcnZ3Rrl07NGjQAP7+/pg0aRJOnjyJEydOvHCXxac5OTnhv//9L4AnScW2bduwcOFCFBQUYP369VK9Tp06YdiwYXj77bcBAOvWrYOW1pO8u3T/165dW2EvmGe/XJ79UX6e9PT0csvc3NwqXMbCwgJFRUW4e/euQtIhhEB6ejrat2+vUgzPo8p5AqjvGJRuu7J1l85PS0srs47bt2+Xia08lpaWkMlkiIuLk77on1Zembro6Ohg5syZmDlzJh48eICDBw9izpw56N27N27evImGDRtCW1sbo0ePxuTJk8tdh4uLC4Anx6KyY/k0AwODMg1+gSc/5E8fM0tLS8jl8nIbTJbOV4W5uTm0tbVx69atSutZWlrCy8sLS5cuLXd+6XeLsiwtLWFhYYF9+/aVO//pqzLKsrCwkP4oeVp5x5vKx4SjDivtoVH6I+Xv749t27Zh586dGDJkiFRv06ZN0vxS//rXv3Djxg38+uuvuHTpEoKCgrBmzRrMmDEDwJMP7NPdcJ/Vo0cPbNu2DY6Ojujfvz8AoGnTpnjllVcwf/58FBYWSldC1K1p06b46KOPsH379jKX6AFg7NixMDIywsiRI/Hw4UN888030NbWRufOndGgQQP88ccfmDJlikZi++677xRuExw/fhzXr1/H+PHjK1zG398fK1aswObNm6XjDwDbt2/Hw4cPFd43Zf7Kfh5VzpOquHDhAn7//XeFWwHff/89TExMnju2SqdOnSCXy7F582YMGzZMKr916xYOHz6scBvo6b8an76SNmDAAHzyySf466+/qnQ5XF0aNGiAoKAg/PXXX5g+fTpSU1PRokUL+Pr64rfffoOXlxf09PQqXN7X1xcrVqwo91g+y9nZGefOnVMoS05OxuXLlxWSiAEDBmDZsmWwsLCQEpsXIZfL0a1bN/z4449YunRphQnLgAEDsGfPHjRu3Fil23MVGTBgALZu3Yri4mJ07NjxhdcHAN26dcOePXsUkrSSkhKFHnml1PE5rIuYcNQRSUlJUjuKzMxMREVFISYmBkOGDJG+OMaMGYN///vfGDt2LFJTU+Hp6Yn4+HgsW7YM/fr1kxKAr7/+Gps3b0Z4eDhatmyJli1bYsqUKfjwww/RuXNndOjQ4bnx+Pv748svv8Tff/+Nzz77TKE8PDwcDRs2LLdLbFWcO3cOU6ZMwbBhw9CkSRPo6enh8OHDOHfuHGbPnl3uMkFBQTA0NERQUBDy8vKwZcsWGBsbY+3atRg7dizu3buHoKAgWFtb4+7du/j9999x9+5drFu37oViPX36NMaPH49hw4bh5s2bmDt3LhwcHDBp0qQKl+nZsyd69+6NDz/8ENnZ2ejcuTPOnTuHBQsW4NVXX8Xo0aOlup6enti6dSt++OEHuLq6wsDAAJ6enirFqOx5UlX29vYYNGgQFi5cCDs7O2zevBkxMTFYvnz5c8cxadCgAebNm4c5c+ZgzJgxeOONN5CZmYlFixbBwMBAoXtz6X4vX74cffv2hba2Nry8vNC5c2e8++67eOutt3D69Gl07doVRkZGSEtLQ3x8PDw9PTFx4sQX2seKDBw4UBozx8rKCtevX8dnn30GJycnNGnSBADw+eefo0uXLvDx8cHEiRPh7OyMnJwcXLlyBbt375ba9EyfPh0bN25E//79sWTJEtjY2OC7777DpUuXymx39OjRGDVqFCZNmoShQ4fi+vXrWLFiRZnbdNOnT8f27dvRtWtXzJgxA15eXigpKcGNGzdw4MABhISEqPwDvnr1anTp0gUdO3bE7Nmz4ebmhjt37uCnn37Chg0bYGJigsWLFyMmJgbe3t6YNm0amjVrhsePHyM1NRV79uzB+vXrn3tb5mmvv/46vvvuO/Tr1w/vvfceOnToAF1dXdy6dQtHjhzB4MGDFZJpZcydOxe7d++Gv78/5s6dC7lcjvXr10vtukqvlALq+RzWSdXYYJXUoLxeKmZmZqJ169Zi9erVUjfAUpmZmeJf//qXsLOzEzo6OsLJyUmEhoZK9c6dOyfkcnmZFu2PHz8Wbdu2Fc7OzgpdOCty//59oaWlJYyMjERBQYFU/t133wkAUpfbpz3dbfVpz7ZAf7aXyp07d0RwcLBwd3cXRkZGwtjYWHh5eYk1a9YodDMrb/1HjhwRxsbGok+fPlJXudjYWNG/f39hbm4udHV1hYODg+jfv7/48ccfpeVKe6k83WK9MqXv04EDB8To0aNFgwYNpN4WKSkpCnXL6xmRl5cnPvzwQ+Hk5CR0dXWFnZ2dmDhxYpn3IjU1VfTq1UuYmJhI3XYrU14vFSGef56UAiAmT56s1DEQ4p/3IDIyUrRs2VLo6ekJZ2dnsXr1aoV6pe/x08f8aV9//bXw8vISenp6wszMTAwePLhMT4b8/Hwxfvx4YWVlJWQyWZmeBBs3bhQdO3YURkZGQi6Xi8aNG4sxY8aI06dPS3Uq6qVS3j5X1BPkaatWrRLe3t7C0tJS6OnpiVdeeUW8/fbbIjU1VaHetWvXxLhx44SDg4PQ1dUVVlZWwtvbW6E3kxBC/PHHH6Jnz57CwMBAmJubi7ffflvs2rWrTC+VkpISsWLFCuHq6ioMDAxEu3btxOHDh8t8toQQIjc3V3z00UeiWbNm0vH19PQUM2bMUOgyrMpx+OOPP8SwYcOEhYWFtN/BwcEK59Pdu3fFtGnThIuLi9DV1RXm5uaibdu2Yu7cuSI3N7fS41refhQWFoqVK1eKVq1aCQMDA2FsbCzc3d3FhAkTFD5zyn7vCCFEXFyc6Nixo9DX1xe2trbi/fffl3pvPXjwQKpX0eewovP62rVrAoBCF/26SCaEEC85xyGqdyIiIvDWW2/h1KlTld6KquucnZ3h4eFR657zU5scPXoUvr6+OHLkSI16vkhd1atXL6SmpiI5Obm6Q6nxeEuFiIhICTNnzsSrr74KR0dH3Lt3D9999x1iYmKkButUOSYcRERESiguLsb8+fORnp4OmUyGFi1a4Ntvv8WoUaOqO7RagbdUiIiISOM48BcRERFpHBMOIiIi0jgmHERERKRxbDSKJ6PF3b59GyYmJioP00xERFSfCSGQk5MDe3t7hQHQnsWEA0+eweDo6FjdYRAREdVaN2/erHREWCYc+OdBPjdv3oSpqWk1R0NERFR7ZGdnw9HR8bkPxWPCgX+eSGlqasqEg4iIqAqe1ySBjUaJiIhI45hwEBERkcbxlgoREdV7xcXFiIuLQ1paGuzs7ODj4wNtbe3qDqtO4RUOIiKq16KiouDm5gZfX1+MHDkSvr6+cHNzQ1RUVHWHVqcw4SAionorKioKQUFB8PT0REJCAnJycpCQkABPT08EBQUx6VAjPrwNT7r0mJmZISsri71UiIjqieLiYri5ucHT0xM7d+5UGLSqpKQEAQEBSEpKQkpKCm+vVELZ31Be4SAionopLi4OqampmDNnTpkRMrW0tBAaGopr164hLi6umiKsW5hwEBFRvZSWlgYA8PDwKHd+aXlpPXoxTDiIiKhesrOzAwAkJSWVO7+0vLQevRgmHEREVC/5+PjA2dkZy5YtQ0lJicK8kpIShIWFwcXFBT4+PtUUYd3ChIOIiOolbW1trFq1CtHR0QgICFDopRIQEIDo6GisXLmSDUbVhAN/ERFRvRUYGIjIyEiEhITA29tbKndxcUFkZCQCAwOrMbq6hd1iwW6xRET1HUcarTplf0N5hYOIiOo9bW1tdO/evbrDqNPYhoOIiIg0jgkHERERaRwTDiIiItI4JhxERESkcUw4iIiISOOYcBAREZHGMeEgIiIijWPCQURERBrHhIOIiIg0jgkHERERaRwTDiIiItI4JhxERESkcUw4iIiISOOYcBAREZHGMeEgIiIijavWhGPdunXw8vKCqakpTE1N0alTJ+zdu1eaL4TAwoULYW9vD7lcju7du+PChQsK68jPz8fUqVNhaWkJIyMjDBo0CLdu3XrZu0JERESVqNaEo1GjRvjkk09w+vRpnD59Gn5+fhg8eLCUVKxYsQKrV6/GF198gVOnTsHW1hY9e/ZETk6OtI7p06djx44d2Lp1K+Lj45Gbm4sBAwaguLi4unaLiIiIniETQojqDuJp5ubm+PTTTzFu3DjY29tj+vTp+PDDDwE8uZphY2OD5cuXY8KECcjKyoKVlRW+/fZbjBgxAgBw+/ZtODo6Ys+ePejdu3e528jPz0d+fr40nZ2dDUdHR2RlZcHU1FTzO0lERFRHZGdnw8zM7Lm/oTWmDUdxcTG2bt2Khw8folOnTrh27RrS09PRq1cvqY6+vj66deuG48ePAwASExNRWFioUMfe3h4eHh5SnfKEhYXBzMxMejk6Ompux4iIiKj6E47z58/D2NgY+vr6+Ne//oUdO3agRYsWSE9PBwDY2Ngo1LexsZHmpaenQ09PDw0bNqywTnlCQ0ORlZUlvW7evKnmvSIiIqKn6VR3AM2aNcPZs2fx4MEDbN++HWPHjkVsbKw0XyaTKdQXQpQpe9bz6ujr60NfX//FAiciojqjuLgYcXFxSEtLg52dHXx8fKCtrV3dYdUp1X6FQ09PD25ubmjXrh3CwsLQqlUrfP7557C1tQWAMlcqMjIypKsetra2KCgowP379yusQ0REVJmoqCi4ubnB19cXI0eOhK+vL9zc3BAVFVXdodUp1Z5wPEsIgfz8fLi4uMDW1hYxMTHSvIKCAsTGxsLb2xsA0LZtW+jq6irUSUtLQ1JSklSHiIioIlFRUQgKCoKnpycSEhKQk5ODhIQEeHp6IigoiEmHGlVrL5U5c+agb9++cHR0RE5ODrZu3YpPPvkE+/btQ8+ePbF8+XKEhYUhPDwcTZo0wbJly3D06FFcvnwZJiYmAICJEyciOjoaERERMDc3x6xZs5CZmYnExESlL4cp28KWiIjqjuLiYri5ucHT0xM7d+6EltY/f4OXlJQgICAASUlJSElJ4e2VSij7G1qtbTju3LmD0aNHIy0tDWZmZvDy8pKSDQD44IMPkJeXh0mTJuH+/fvo2LEjDhw4ICUbALBmzRro6Ohg+PDhyMvLg7+/PyIiInhyEBFRpeLi4pCamootW7YoJBsAoKWlhdDQUHh7eyMuLg7du3evniDrkBo3Dkd14BUOIqL6Z8uWLRg5ciRycnJgbGxcZn5OTg5MTU3x/fff44033qiGCGuHWjcOBxER0ctkZ2cHAEhKSip3fml5aT16MUw4iIioXvLx8YGzszOWLVuGkpIShXklJSUICwuDi4sLfHx8qinCuoUJBxER1Uva2tpYtWoVoqOjERAQoNBLJSAgANHR0Vi5ciXbBKpJtQ/8RUREVF0CAwMRGRmJkJAQheEUXFxcEBkZicDAwGqMrm5ho1Gw0SgRUX3HkUarrlZ0iyUiIqoJtLW12fVVw9iGg4iIiDSOCQcRERFpHBMOIiIi0jgmHERERKRxTDiIiIhI45hwEBERkcYx4SAiIiKNY8JBREREGseEg4iIiDSOCQcRERFpnMpDmxcXFyMiIgKHDh1CRkZGmUf6Hj58WG3BERERUd2gcsLx3nvvISIiAv3794eHhwdkMpkm4iIiIqI6ROWEY+vWrdi2bRv69euniXiIiIioDlK5DYeenh7c3Nw0EQsRERHVUSonHCEhIfj8888hhNBEPERERFQHKXVLJTAwUGH68OHD2Lt3L1q2bAldXV2FeVFRUeqLjoiIiOoEpRIOMzMzhekhQ4ZoJBgiIiKqm5RKOMLDwzUdBxFRGcXFxYiLi0NaWhrs7Ozg4+MDbW3t6g6LiKpA5TYcfn5+ePDgQZny7Oxs+Pn5qSMmIiJERUXBzc0Nvr6+GDlyJHx9feHm5sbbtkS1lMoJx9GjR1FQUFCm/PHjx4iLi1NLUERUv0VFRSEoKAienp5ISEhATk4OEhIS4OnpiaCgICYdRLWQTCjZ3eTcuXMAgNatW+Pw4cMwNzeX5hUXF2Pfvn3YsGEDUlNTNRKoJmVnZ8PMzAxZWVkwNTWt7nCI6rXi4mK4ubnB09MT27dvx7Fjx6RbKp07d8bQoUORlJSElJQU3l4hqgGU/Q1VeuCv1q1bQyaTQSaTlXvrRC6XY+3atVWLlojof+Li4pCamooJEyagadOmCn/EODs7491338Xu3bsRFxeH7t27V1ucRKQapROOa9euQQgBV1dX/Prrr7CyspLm6enpwdramn9tENELS0tLAwDMmTMHAwYMwJYtW+Dh4YGkpCQsW7YMc+fOVahHRLWD0gmHk5MTAJR5WBsRkTpZW1sDADp37oydO3dCS+tJU7PXXnsNO3fuRLdu3RAfHy/VI6LaQeVnqfz000/llstkMhgYGMDNzQ0uLi4vHBgRUXk4yjFR7aRywhEQEACZTFbmQ19aJpPJ0KVLF+zcuRMNGzZUW6BEVD9kZGQAAOLj4xEQEIDQ0FDplkpYWBiOHTumUI+IageVu8XGxMSgffv2iImJQVZWFrKyshATE4MOHTogOjoav/zyCzIzMzFr1ixNxEtEdZydnR0AICwsDOfPn4e3tzdMTU3h7e0tteN4uh4R1Q5Kd4st5eHhga+++gre3t4K5ceOHcO7776LCxcu4ODBgxg3bhxu3Lih1mA1hd1iiWoOdoslql2U/Q1V+QrH1atXy12hqakp/vzzTwBAkyZN8Pfff6u6aiIiaGtrY9WqVYiOjsbQoUOhr6+PAQMGQF9fH0OHDkV0dDRWrlzJZIOollE54Wjbti3ef/993L17Vyq7e/cuPvjgA7Rv3x4AkJKSgkaNGqkvSiKqVwIDAxEZGVnuLZXIyMgyT7AmoppP5Uaj//3vfzF48GA0atQIjo6OkMlkuHHjBlxdXbFr1y4AQG5uLubNm6f2YImofnn2ji+75RPVXiq34QCefAns378fycnJEELA3d0dPXv2lPrL1zZsw0FUs5Q+S2XAgAGYM2eOwsBf0dHRvMpBVIMo+xtapYSjrmHCQVRzPN1o9OmBv4AnVzgCAgLYaJSoBlH7s1SedujQIRw6dAgZGRllLnFu3LixKqskIgLwz7NUtmzZUuaqqZaWFkJDQ+Ht7c1nqRDVMionHIsWLcLixYvRrl072NnZQSaTaSIuIqqnSp+R4uHhUe780nI+S4WodlG50cX69esRERGBkydPYufOndixY4fCSxVhYWFo3749TExMYG1tjYCAAFy+fFmhTnBwsPSU2tLXa6+9plAnPz8fU6dOhaWlJYyMjDBo0CDcunVL1V0johqgdECvpKSkcueXlnPgL6LaReWEo6CgoMygX1UVGxuLyZMn48SJE4iJiUFRURF69eqFhw8fKtTr06cP0tLSpNeePXsU5k+fPh07duzA1q1bER8fj9zcXAwYMADFxcVqiZOIXh4fHx84Oztj2bJlZW7ZlpSUICwsDC4uLvDx8ammCImoSoSKPvjgA7F48WJVF1NKRkaGACBiY2OlsrFjx4rBgwdXuMyDBw+Erq6u2Lp1q1T2119/CS0tLbFv3z6ltpuVlSUAiKysrCrHTkTqs337diGTycTAgQPF8ePHRXZ2tjh+/LgYOHCgkMlkYvv27dUdIhH9j7K/oSq34Xj8+DG++uorHDx4EF5eXtDV1VWYv3r16ionP1lZWQAAc3NzhfKjR4/C2toaDRo0QLdu3bB06VLp0dSJiYkoLCxEr169pPr29vbw8PDA8ePH0bt37zLbyc/PR35+vjSdnZ1d5ZiJSP1KB/4KCQlRuKLq4uLCLrFEtZTKCce5c+fQunVrAGXvsb5IA1IhBGbOnIkuXbooNBbr27cvhg0bBicnJ1y7dg3z5s2Dn58fEhMToa+vj/T0dOjp6ZV5Mq2NjQ3S09PL3VZYWBgWLVpU5ViJSPMCAwMxePBgxMXFSc9S8fHxYVdYolpK5YTjyJEjmogDU6ZMwblz5xAfH69QPmLECOn/Hh4eaNeuHZycnPDzzz9X+leOEKLCBCg0NBQzZ86UprOzs+Ho6PiCe0BE6qatrc2ur0R1RJWHBr1y5Qr279+PvLw8AGWHIFbF1KlT8dNPP+HIkSPPfQaLnZ0dnJyckJKSAgCwtbVFQUEB7t+/r1AvIyMDNjY25a5DX18fpqamCi8iIiLSHJUTjszMTPj7+6Np06bo16+f1Bd+/PjxCAkJUWldQghMmTIFUVFROHz4MFxcXJTa/s2bN6UucW3btoWuri5iYmKkOmlpaUhKSlJbbxoiIiJ6MSonHDNmzICuri5u3LgBQ0NDqXzEiBHYt2+fSuuaPHkyNm/ejO+//x4mJiZIT09Henq6dNUkNzcXs2bNQkJCAlJTU3H06FEMHDgQlpaWGDJkCADAzMwMb7/9NkJCQnDo0CH89ttvGDVqFDw9PdGjRw9Vd4+IiIg0QOU2HAcOHMD+/fvL3Ppo0qQJrl+/rtK61q1bBwBl7tGGh4cjODgY2traOH/+PDZt2oQHDx7Azs4Ovr6++OGHH2BiYiLVX7NmDXR0dDB8+HDk5eXB398fERERbFxGRERUQ6iccDx8+FDhykapv//+G/r6+iqt63ntPuRyOfbv3//c9RgYGGDt2rVYu3atStsnIiKil0PlWypdu3bFpk2bpGmZTIaSkhJ8+umn8PX1VWtwREREVDeofIXj008/Rffu3XH69GkUFBTggw8+wIULF3Dv3j0cO3ZMEzESERFRLadywtGiRQucO3cO69atg7a2Nh4+fIjAwEBMnjyZD1MiIqJaqbi4mIPMaZhMvMgAGnVEdnY2zMzMkJWVxTE5iIjqmaioKMycOVOh44OTkxNWr17NYfSVoOxvqFJXOM6dO6f0hr28vJSuS0REVJ2ioqIwdOhQyOVyhfKMjAwMHToU27dvZ9KhJkpd4dDS0oJMJnturxKZTFYrHwnPKxxERPVPcXEx7OzscPfuXQwYMABz586Fh4cHkpKSsHTpUkRHR8Pa2hq3b9/m7ZVKqPUKx7Vr19QWGBERUU1w9OhR3L17F126dMGuXbugpfWk4+Zrr72GXbt2oVu3boiPj8fRo0fh7+9fzdHWfkolHE5OTpqOg4iI6KU6evQoAGDRokVSslFKS0sLCxYsQM+ePZlwqInKvVSIiIhqi0ePHuHSpUvlzit9FlhKSgoaNGiAvLw8pKamwtnZGXK5XHpIaFpaGs6cOVPhNtzd3csdEJMUsZcK2IaDiKiuOnPmDNq2bavRbSQmJqJNmzYa3UZNptY2HERERLWRu7s7EhMTy51XXFyM3r174/79+/Dx8YGfnx8WLVqEBQsW4PDhw4iLi4O5uTn27dtXaaNRd3d3TYVfp/AKB3iFg4iovoqKikJQUBAMDAykJ5UDgKGhIfLy8hAZGclusc+h7G+oys9SAYAHDx7g66+/RmhoKO7duwfgyWWrv/76q2rREhERVYPAwEBERkbCxsZGodzGxobJhpqpnHCcO3cOTZs2xfLly7Fy5Uo8ePAAALBjxw6EhoaqOz4iIiKNCgwMxJUrV7BhwwYAwIYNG5CSksJkQ81UTjhmzpyJ4OBgpKSkwMDAQCrv27cvfvnlF7UGR0RE9DJoa2ujXbt2AIB27dpxoC8NULnR6KlTp6Qs8GkODg5IT09XS1BEREQVSUlJQU5OjtrXe/HiRYV/NcHExARNmjTR2PprMpUTDgMDA2RnZ5cpv3z5MqysrNQSFBERUXlSUlLQtGlTjW5j1KhRGl1/cnJyvUw6VE44Bg8ejMWLF2Pbtm0Anjw/5caNG5g9ezaGDh2q9gCJiIhKlV7Z2Lx5M5o3b67WdT878Je6Xbx4EaNGjdLI1ZnaQOWEY+XKlejXrx+sra2Rl5eHbt26IT09HZ06dcLSpUs1ESMREZHE1liGNnbaaG5bpY6WlTBCZ5eWal7nP+QPtGFrLNPY+ms6lRMOU1NTxMfH4/Dhwzhz5gxKSkrQpk0b9OjRQxPxERERKZjQVg/Nf5kA1LJ+Cs3xJPb6qsojjfr5+cHPz0+dsRARET3XhsQCjJgfgea1bITPi5cuYcOqkRhU3YFUE5UTjmnTpsHNzQ3Tpk1TKP/iiy9w5coVfPbZZ+qKjYiISMGjR4+Qnitw7M9c5DUoUeu6Nd6GI60Y6bn1d3BvlROO7du346effipT7u3tjU8++YQJBxERaUzpk1/feeedao6k6kxMTKo7hGqhcsKRmZkJMzOzMuWmpqb4+++/1RIUERFReQICAgBo5pHwpb1INNEDphTH4VCBm5sb9u3bhylTpiiU7927F66urmoLjIiI6FmWlpYYP368RrfRvHnzev24eU1ROeGYOXMmpkyZgrt370qNRg8dOoRVq1bxdgoRERGVS+WEY9y4ccjPz8fSpUvx8ccfAwCcnZ2xbt06jBkzRu0BEhERUe1XpW6xEydOxMSJE3H37l3I5XIYGxurOy4iqsMePXokNf57nqr2HNDEPX4iqroqj8MBgM9OIaIquXTpEtq2bavRbSQmJvI+PKmU3Fb14W1MbpWjcsJx584dzJo1C4cOHUJGRgaEUOxTXFxcrLbgiKhucnd3R2JiolJ1q9pzwL2WDQpFmlGV5FbVh7cxuVWOyglHcHAwbty4gXnz5sHOzg4yWf0dF56IqsbQ0FDlL2j2HKCqUCW5fZHbd/R8Kicc8fHxiIuLQ+vWrTUQDhERkfqomtx27txZg9HUbyo/as/R0bHMbRQiIiKiyqh8heOzzz7D7NmzsWHDBjg7O2sgJCKqjVJSUpCTk6P29Va1IZ8q6vPoj0Qvi0yoeLmiYcOGePToEYqKimBoaAhdXV2F+ffu3VNrgC9DdnY2zMzMkJWVBVNT0+oOh6jWSUlJQdOmTas7jBeSnJzMpIOoCpT9Da3SFQ4ioqeVXtnQxDMoNP4Ez//1gtHE1Rki+ofKCcfYsWM1EQcR1QGa6knChnxEtZ/KjUYB4OrVq/joo4/wxhtvICMjAwCwb98+XLhwQa3BERERUd2gcsIRGxsLT09PnDx5ElFRUcjNzQUAnDt3DgsWLFB7gERERFT7qZxwzJ49G0uWLEFMTAz09PSkcl9fXyQkJKg1OCIiIqobVE44zp8/jyFDhpQpt7KyQmZmplqCIiIiorpF5YSjQYMGSEtLK1P+22+/wcHBQS1BERERUd2icsIxcuRIfPjhh0hPT4dMJkNJSQmOHTuGWbNmYcyYMSqtKywsDO3bt4eJiQmsra0REBCAy5cvK9QRQmDhwoWwt7eHXC5H9+7dyzROzc/Px9SpU2FpaQkjIyMMGjQIt27dUnXXiIiISENUTjiWLl2KV155BQ4ODsjNzUWLFi3QtWtXeHt746OPPlJpXbGxsZg8eTJOnDiBmJgYFBUVoVevXnj48KFUZ8WKFVi9ejW++OILnDp1Cra2tujZs6dCn/np06djx44d2Lp1K+Lj45Gbm4sBAwbwybVEREQ1haiiK1euiB9//FH88MMPIjk5uaqrUZCRkSEAiNjYWCGEECUlJcLW1lZ88sknUp3Hjx8LMzMzsX79eiGEEA8ePBC6urpi69atUp2//vpLaGlpiX379im13aysLAFAZGVlqWU/iOqbxMREAUAkJiZWdygqq82xE9UEyv6GqjzwV6nGjRujcePGakp7nsjKygIAmJubAwCuXbuG9PR09OrVS6qjr6+Pbt264fjx45gwYQISExNRWFioUMfe3h4eHh44fvw4evfuXWY7+fn5yM/Pl6azs7PVuh9ERESkSOWEY9y4cZXO37hxY5UCEUJg5syZ6NKlCzw8PAAA6enpAAAbGxuFujY2Nrh+/bpUR09PDw0bNixTp3T5Z4WFhWHRokVVipOIiIhUp3LCcf/+fYXpwsJCJCUl4cGDB/Dz86tyIFOmTMG5c+cQHx9fZp5MJlOYFkKUKXtWZXVCQ0Mxc+ZMaTo7OxuOjo5ViJqIiIiUoXLCsWPHjjJlJSUlmDRpElxdXasUxNSpU/HTTz/hl19+QaNGjaRyW1tbAE+uYtjZ2UnlGRkZ0lUPW1tbFBQU4P79+wpXOTIyMuDt7V3u9vT19aGvr1+lWImIiEh1VXqWSpmVaGlhxowZWLNmjUrLCSEwZcoUREVF4fDhw3BxcVGY7+LiAltbW8TExEhlBQUFiI2NlZKJtm3bQldXV6FOWloakpKSKkw4iIiI6OWqcqPRZ129ehVFRUUqLTN58mR8//332LVrF0xMTKQ2F2ZmZpDL5ZDJZJg+fTqWLVuGJk2aoEmTJli2bBkMDQ0xcuRIqe7bb7+NkJAQWFhYwNzcHLNmzYKnpyd69Oihrt0jIiKiF6BywvF02wfgyVWKtLQ0/Pzzzyo/un7dunUAgO7duyuUh4eHIzg4GADwwQcfIC8vD5MmTcL9+/fRsWNHHDhwACYmJlL9NWvWQEdHB8OHD0deXh78/f0REREBbW1tVXePiKrI1lgG+YNk4LZaLpy+NPIHybA1rrxNGBG9OJkQQqiygK+vr8K0lpYWrKys4Ofnh3HjxkFHR20XTV6a7OxsmJmZISsrC6amptUdDlGtc+bMGfw00xsLu9fOtlELj+Zj0OrjaNOmTXWHQlTrKPsbqnJ2cOTIkRcKjIjqpg2JBRgxPwLN3d2rOxSVXLx0CRtWjcSg6g6EqI6rfZcjiKhGSs8VyGvQFLBvXd2hqCQvvQTpuSpd6CWiKlA54Xj11VefOwZGqTNnzqgcEBEREdU9Kiccffr0wZdffokWLVqgU6dOAIATJ07gwoULmDhxIuRyudqDJCIiotpN5YTj7t27mDZtGj7++GOF8gULFuDmzZtVHtqciIiI6i6V+6/9+OOPGDNmTJnyUaNGYfv27WoJioiIiOoWla9wyOVyxMfHo0mTJgrl8fHxMDAwUFtgRFR7PHr0CIBm2m3l5eUhNTUVzs7OGrlle/HiRbWvk4jKUjnhmD59OiZOnIjExES89tprAJ604di4cSPmz5+v9gCJqOa7dOkSAOCdd96p5kiq7unBBIlI/VROOGbPng1XV1d8/vnn+P777wEAzZs3R0REBIYPH672AImo5gsICAAAuLu7w9DQUK3rvnjxIkaNGoXNmzejefPmal13KRMTkzJXbYlIvVQeabQu4kijRDXXmTNn0LZtWyQmJnIkUKIaSNnf0Co99ODBgwf4+uuvMWfOHNy7dw/Aky+Fv/76q2rREhERUZ2m8i2Vc+fOoUePHjAzM0NqairGjx8Pc3Nz7NixA9evX8emTZs0EScRERHVYipf4Zg5cyaCg4ORkpKi0Culb9+++OWXX9QaHBEREdUNKiccp06dwoQJE8qUOzg4ID09XS1BERERUd2icsJhYGCA7OzsMuWXL1+GlZWVWoIiIiKiukXlhGPw4MFYvHgxCgsLAQAymQw3btzA7NmzMXToULUHSERERLWfygnHypUrcffuXVhbWyMvLw/dunWDm5sbTExMsHTpUk3ESERERLWcyr1UTE1NER8fj8OHD+PMmTMoKSlBmzZt0KNHD03ER0RERHWAyglHKT8/P/j5+akzFiIiIqqjlL6lcvLkSezdu1ehbNOmTXBxcYG1tTXeffdd5Ofnqz1AIiIiqv2UTjgWLlyIc+fOSdPnz5/H22+/jR49emD27NnYvXs3wsLCNBIkERER1W5KJxxnz56Fv7+/NL1161Z07NgR//nPfzBz5kz83//9H7Zt26aRIImIiKh2UzrhuH//PmxsbKTp2NhY9OnTR5pu3749bt68qd7oiIiIqE5QOuGwsbHBtWvXAAAFBQU4c+YMOnXqJM3PycmBrq6u+iMkIiKiWk/phKNPnz6YPXs24uLiEBoaCkNDQ/j4+Ejzz507h8aNG2skSCIiIqrdlO4Wu2TJEgQGBqJbt24wNjbGN998Az09PWn+xo0b0atXL40ESURERLWb0gmHlZUV4uLikJWVBWNjY2hrayvM//HHH2FsbKz2AImIiKj2U3ngLzMzs3LLzc3NXzgYqh2Ki4sRFxeHtLQ02NnZwcfHp0wCSlSZR48e4dKlS0rVvXjxosK/ynJ3d4ehoaHKsRGRZlR5pFGqn6KiohASEoLU1FSpzNnZGatWrUJgYGD1BUa1yqVLl9C2bVuVlhk1apRK9RMTE9GmTRuVliEizWHCQUqLiopCUFAQBgwYgC1btsDDwwNJSUlYtmwZgoKCEBkZyaSDlOLu7o7ExESl6ubl5SE1NRXOzs6Qy+UqbYOIag6ZEEJUdxDVLTs7G2ZmZsjKyoKpqWl1h1MjFRcXw83NDZ6enti5cye0tP7p4FRSUoKAgAAkJSUhJSWFt1eIiOoRZX9DVX48PdVPcXFxSE1NxZw5cyCEwNGjR7FlyxYcPXoUQgiEhobi2rVriIuLq+5QiYioBuItFVJKWloaAODq1at44403yrThWLJkiUI9IiKip/EKBynFzs4OwJOGe56enkhISEBOTg4SEhLg6ekpNegrrUdERPQ0tuEA23Aoo6CgAEZGRrCwsMCtW7ego/PPxbGioiI0atQImZmZePjwocKAcEREVLexDQep1fHjx1FUVISMjAwEBgYqXOEIDAxERkYGioqKcPz48eoOlYiIaiC24SBJZYMxlSYSH3/8Mb788kt4e3tL8xwcHLB48WLMmzcPx48frzTD5WBMRET1E2+pgLdUSp05c0blwZhUxcGYiIjqFmV/Q3mFgySVDcZUXFyMgIAAuLm5YdWqVbh8+TJGjRqFzZs3o1mzZggJCcHVq1exY8eOSsfh4GBMRET1ExMOkhgaGlZ69WHt2rUICgrC4sWLpRFFCwsLsXjxYsTFxSEyMhLt27d/WeESEVEtwlsq4C0VVZT3LBUXFxesXLmSw5oTEdVD7KVCGhEYGIgrV65gw4YNAIANGzYgJSWFyQYREVWqWhOOX375BQMHDoS9vT1kMhl27typMD84OBgymUzh9dprrynUyc/Px9SpU2FpaQkjIyMMGjQIt27deol7Uf9oa2ujXbt2AIB27drx2SlERPRc1ZpwPHz4EK1atcIXX3xRYZ0+ffogLS1Neu3Zs0dh/vTp07Fjxw5s3boV8fHxyM3NxYABA1BcXKzp8ImIiEhJ1dpotG/fvujbt2+ldfT19WFra1vuvKysLPz3v//Ft99+ix49egAANm/eDEdHRxw8eBC9e/dWe8xE9PIUFxcjLi4OaWlpsLOzg4+PD6+oEdVSNb4Nx9GjR2FtbY2mTZvinXfeQUZGhjQvMTERhYWF6NWrl1Rmb28PDw+PSke8zM/PR3Z2tsKLiGqWqKgouLm5wdfXFyNHjoSvry/c3NwQFRVV3aERURXU6ISjb9+++O6773D48GGsWrUKp06dgp+fH/Lz8wEA6enp0NPTQ8OGDRWWs7GxQXp6eoXrDQsLg5mZmfRydHTU6H4QkWqioqIQFBRU7oMCg4KCmHQQ1UI1ehyOESNGSP/38PBAu3bt4OTkhJ9//rnSXhFCCMhksgrnh4aGYubMmdJ0dnZ2nU06UlJSkJOTo/b1Xrx4UeFfTTAxMUGTJk00tn6qmYqLixESEoIBAwZg586d0NJ68nfRa6+9hp07dyIgIACzZs3C4MGDeXuFqBap0QnHs+zs7ODk5ISUlBQAgK2tLQoKCnD//n2FqxwZGRkKz/p4lr6+PvT19TUeb3VLSUlB1zbNYGdccfL1Il611cKqWWM0sm4ASMsV+OXMZSYd9UxcXBxSU1OxZcsWKdkopaWlhdDQUHh7eyMuLg7du3evniCJSGW1KuHIzMzEzZs3YWdnBwBo27YtdHV1ERMTg+HDhwMA0tLSkJSUhBUrVlRnqDVCTk4OJrTVw8LutTO5Wng0XyNXZ6hmS0tLA/DkqmZ5SstL6xFR7VCtCUdubi6uXLkiTV+7dg1nz56Fubk5zM3NsXDhQgwdOhR2dnZITU3FnDlzYGlpiSFDhgAAzMzM8PbbbyMkJAQWFhYwNzfHrFmz4OnpKfVaqe82JBZgxPwINK9lzzC5eOkSNqwaiUHVHQi9dKV/UCQlJZUZd6e0/Ol6RFQ7VGvCcfr0afj6+krTpe0qxo4di3Xr1uH8+fPYtGkTHjx4ADs7O/j6+uKHH36AiYmJtMyaNWugo6OD4cOHIy8vD/7+/oiIiOC93f9JzxXIa9AUsG9d3aGoJC+9BOm59X7U/XrJx8cHzs7OWLZsmUIbDgAoKSlBWFgYXFxc4OPjU41REpGqqjXh6N69Oyp7lMv+/fufuw4DAwOsXbsWa9euVWdoRFRNtLW1sWrVKgQFBSEgIAChoaHw8PBAUlISwsLCEB0djcjISP5RQVTL1Ko2HERUPwQGBiIyMhIhISEKDcBdXFwQGRnJZ/cQ1UJMOOqwR48eAQDOnDmj9nXn5eUhNTUVzs7OkMvlal+/JrvbUu0QGBiIwYMHc6RRojqCCUcddunSJQDAO++8U82RVN3T7XWo/tHW1mbXV6I6gglHHRYQEAAAcHd3h6GhoVrXffHiRYwaNQqbN29G8+bN1bruUhz4i4io7mDCUYdZWlpi/PjxGt1G8+bN0aZNG41ug4iIaj8mHCR59OiRdBvmeao6tLkmrrYQEVHNx4SDJJcuXULbtm1VWmbUqFEq1U9MTOQVESKieogJB0nc3d2RmJioVN2q9lJxr2UjnhIRkXrIRGUjb9UT2dnZMDMzQ1ZWFkxNTas7HCIiolpD2d9QrQrnEFXg3r178PT0hIWFBTw9PXHv3r3qDomIiGo43lIhldja2uLOnTvS9L1792BhYQEbGxukp6dXY2RERFST8QoHKe3ZZONpd+7cga2t7UuOiIiIagsmHKSUe/fuScmGjo4OZs+ejStXrmD27NnQ0XlyoezOnTu8vUJEROVio1Gw0agyWrRoIY25kZ+fDz09PWleQUEB9PX1ATwZCOyPP/6olhiJiOjlY6NRUqs///wTAPD6669DW1sbR48exZYtW3D06FFoa2tj2LBhCvWIiIiexkajpJTSJ3SeOHECbm5uSE1NleY5OzujpKREoR4REdHTeIWDlDJ8+HAAQGpqKtzd3ZGQkICcnBwkJCTA3d0dN27cUKhHRET0NLbhANtwKCM3N1fhUfEdOnTA4sWLMX/+fPz6669SeU5ODoyNjasjRCIiqgZsw0Fqdfr0aYXpX3/9FX369FFINsqrR0REBDDhICWlpaUBAPr371/u/NLy0npERERPY8JBSrGzswMAfPTRR3j06BEmT56MXr16YfLkyXj06BHmzp2rUI+IiOhpbMMBtuFQRnFxMdzc3ODp6YmdO3dCS+ufXLWkpAQBAQFISkpCSkoKe6oQEdUjbMNBaqWtrY1Vq1YhOjoaAQEBCr1UAgICEB0djZUrVzLZICKicnEcDlJaYGAgIiMjERISAm9vb6ncxcUFkZGRCAwMrMboiIioJuMtFfCWiqqKi4sRFxeHtLQ02NnZwcfHh1c2iIjqKWV/Q3mFg1Smra2N7t27V3cYRERUi7ANBxEREWkcEw4iIiLSOCYcREREpHFswwGgtN1sdnZ2NUdCRERUu5T+dj6vDwoTDjx54BgAODo6VnMkREREtVNOTg7MzMwqnM9usXgyUubt27dhYmICmUxW3eHUCtnZ2XB0dMTNmzfZlZg0iucavSw816pGCIGcnBzY29srjEL9LF7hAKClpYVGjRpVdxi1kqmpKT+Y9FLwXKOXheea6iq7slGKjUaJiIhI45hwEBERkcYx4aAq0dfXx4IFC6Cvr1/doVAdx3ONXhaea5rFRqNERESkcbzCQURERBrHhIOIiIg0jgkHERERaRwTDiIiolpAJpNh586d1R1GlTHhqKGCg4MREBBQ3WHUCunp6Zg6dSpcXV2hr68PR0dHDBw4EIcOHaru0GoFnmvKS09Px3vvvQc3NzcYGBjAxsYGXbp0wfr16/Ho0aPqDq/G4bmlvPrwPcaRRqlWS01NRefOndGgQQOsWLECXl5eKCwsxP79+zF58mRcunSpukOkOuLPP/+UzrVly5bB09MTRUVFSE5OxsaNG2Fvb49BgwZVd5hUC9Wb7zFBNdLYsWPF4MGDpemSkhLRuHFj8emnnyrUO3/+vJDJZOLKlStCCCEWLFggHB0dhZ6enrCzsxNTp06V6t67d0+MHj1aNGjQQMjlctGnTx+RnJwsze/WrZsAUOZ17do18dZbb4n+/fsrbLuwsFDY2NiI//73v9LyU6ZMEe+9955o0KCBsLa2Fhs2bBC5ubkiODhYGBsbC1dXV7Fnzx6F9Vy4cEH07dtXGBkZCWtrazFq1Chx9+5dpY5T3759hYODg8jNzS0z7/79+0qto77juabcuda7d2/RqFGjcs+10uNGinhuqfd7DID4z3/+IwICAoRcLhdubm5i165dSm2jJmDCUUM9+0EVQoilS5eKFi1aKJTNmDFDdO3aVQghxI8//ihMTU3Fnj17xPXr18XJkyfFV199JdUdNGiQaN68ufjll1/E2bNnRe/evYWbm5soKCgQQgiRmZkp0tLSpFdgYKBo1qyZePTokTh27JjQ1tYWt2/flta3a9cuYWRkJHJycoQQTz6oJiYm4uOPPxbJycni448/FlpaWqJv377iq6++EsnJyWLixInCwsJCPHz4UAghxO3bt4WlpaUIDQ0VFy9eFGfOnBE9e/YUvr6+zz1GmZmZQiaTiWXLlql+gEnCc+3559rff/8tZDKZCAsLU/0A12M8t9T7PQZANGrUSHz//fciJSVFTJs2TRgbG4vMzMznLlsTMOGoocr7oN6+fVtoa2uLkydPCiGEKCgoEFZWViIiIkIIIcSqVatE06ZNpQ/e05KTkwUAcezYMans77//FnK5XGzbtq1M/dWrV4sGDRqIy5cvS2UtWrQQy5cvl6YDAgJEcHCwNN2tWzfRpUsXabqoqEgYGRmJ0aNHS2VpaWkCgEhISBBCCDFv3jzRq1cvhW3fvHlTAFDYdnlOnjwpAIioqKhK61HleK49/1w7ceJEueeahYWFMDIyEkZGRuKDDz6odB31Ec8t9X6PARAfffSRNJ2bmytkMpnYu3fvc5etCdhotBaxs7ND//79sXHjRgBAdHQ0Hj9+jGHDhgEAhg0bhry8PLi6uuKdd97Bjh07UFRUBAC4ePEidHR00LFjR2l9FhYWaNasGS5evKiwnb1792L27Nn44Ycf0LRpU6l8/PjxCA8PBwBkZGTg559/xrhx4xSW9fLykv6vra0NCwsLeHp6SmU2NjbS8gCQmJiII0eOwNjYWHq5u7sDAK5evVrp8RD/GyRXJpNVWo9Ux3OtfM+ea7/++ivOnj2Lli1bIj8/X6l11Hc8txSp+j32dGxGRkYwMTGR4qjpmHDUMuPHj8fWrVuRl5eH8PBwjBgxAoaGhgAAR0dHXL58Gf/+978hl8sxadIkdO3aFYWFhdJJ/SwhhMKJ/scff+D111/HJ598gl69einUHTNmDP78808kJCRg8+bNcHZ2ho+Pj0IdXV1dhWmZTKZQVrqtkpIS6d+BAwfi7NmzCq+UlBR07dq10mPRpEkTyGSyMl80pB481/7h5uYGmUxWpvGeq6sr3NzcIJfLK12eFPHc+oeq32PlxVYaR03HXiq1TL9+/WBkZIR169Zh7969+OWXXxTmy+VyDBo0CIMGDcLkyZPh7u6O8+fPo0WLFigqKsLJkyfh7e0NAMjMzERycjKaN28uTQ8cOBCBgYGYMWNGmW1bWFggICAA4eHhSEhIwFtvvfXC+9OmTRts374dzs7O0NFR7XQ0NzdH79698e9//xvTpk2DkZGRwvwHDx6gQYMGLxxjfcVzTTGenj174osvvsDUqVPLnGukGp5b/6hP32NMOGqwrKwsnD17VqHM3NwcwcHBCA0NhZubGzp16iTNi4iIQHFxMTp27AhDQ0N8++23kMvlcHJygoWFBQYPHox33nkHGzZsgImJCWbPng0HBwcMHjwYABAYGAi5XI6FCxciPT1dWq+VlRW0tbUBPPnLZMCAASguLsbYsWNfeB8nT56M//znP3jjjTfw/vvvw9LSEleuXMHWrVvxn//8R9puRb788kt4e3ujQ4cOWLx4Mby8vFBUVISYmBisW7eOVz+UxHNNuXOtc+fOaNeuHRYuXAgvLy9oaWnh1KlTuHTpEtq2bfvCMdZFPLf4PSapvuYjVJmxY8eW27Vr7Nix4urVqwKAWLFihcIyO3bsEB07dhSmpqbCyMhIvPbaa+LgwYPS/NLuZGZmZkIul4vevXsrdCcrb3v4X3eyUiUlJcLJyUn069evTMzdunUT7733nkKZk5OTWLNmjUIZALFjxw5pOjk5WQwZMkTq5ubu7i6mT5+udDfD27dvi8mTJwsnJyehp6cnHBwcxKBBg8SRI0eUWr6+47mm2rk2ZcoU4eLiInR1dYWxsbHo0KGD+PTTT6UeC/QPnlvq/R57dptCCGFmZibCw8OV2kZ14+Ppa6Fjx46he/fuuHXrltR46WV59OgR7O3tsXHjRgQGBr7UbdPLx3ONNIXnVv3DWyq1SH5+Pm7evIl58+Zh+PDhL/VDWlJSgvT0dKxatQpmZmYcUbGO47lGmsJzq/5iL5VaZMuWLWjWrBmysrKwYsWKl7rtGzduwMHBAdu2bcPGjRtVbhhV1W0+3c3s2deNGzc0HkN9xXON55qm8Nyqv+cWb6lQjVVUVITU1NQK51elRThReXiukabw3PoHEw4iIiLSON5SISIiIo1jwkFEREQax4SDiIiINI4JBxEREWkcEw4i0riFCxeidevWL7yer776Co6OjtDS0sJnn332wuurSEREhMLzK56NPzg4GAEBARrbPlFdxISDqA4KDg6GTCaTnnLp6uqKWbNm4eHDhy+03qomDrNmzcKhQ4deaNvZ2dmYMmUKPvzwQ/z111949913X2h9lRkxYgSSk5MrnP/5558jIiJCY9snqovqR+dfonqoT58+CA8PR2FhIeLi4jB+/Hg8fPgQ69atK1O3sLCwzGOv1al0kKMXcePGDRQWFqJ///6ws7NTU2Tlk8vllT5y3szMTKPbJ6qLeIWDqI7S19eHra0tHB0dMXLkSLz55pvYuXMngH+uVGzcuBGurq7Q19eHEAI3btzA4MGDYWxsDFNTUwwfPhx37twB8OQ2w6JFi/D7779LV09K/8rPysrCu+++C2tra5iamsLPzw+///67FEtFtyRWrlwJOzs7WFhYYPLkySgsLCx3XyIiIuDp6QkAcHV1hUwmQ2pqKq5evYrBgwfDxsYGxsbGaN++PQ4ePKiwrLOzM5YsWYIxY8bA2NgYTk5O2LVrF+7evSvtq6enJ06fPq2wvcoeCf7sLRUhBFasWAFXV1fI5XK0atUKkZGR0vz79+/jzTffhJWVFeRyOZo0aYLw8PAK109UFzHhIKon5HK5wg/6lStXsG3bNmzfvl16fHhAQADu3buH2NhYxMTE4OrVqxgxYgSAJ7cZQkJC0LJlS6SlpSEtLQ0jRoyAEAL9+/dHeno69uzZg8TERLRp0wb+/v64d+9ehfEcOXIEV69exZEjR/DNN98gIiKiwtsUI0aMkBKJX3/9FWlpaXB0dERubi769euHgwcP4rfffkPv3r0xcODAMsNFr1mzBp07d8Zvv/2G/v37Y/To0RgzZgxGjRqFM2fOwM3NDWPGjEFVx0H86KOPEB4ejnXr1uHChQuYMWMGRo0ahdjYWADAvHnz8Mcff2Dv3r24ePEi1q1bB0tLyypti6jWqq7H1BKR5owdO1YMHjxYmj558qSwsLAQw4cPF0IIsWDBAqGrqysyMjKkOgcOHBDa2trixo0bUtmFCxcEAPHrr79Ky7Vq1UphW4cOHRKmpqbi8ePHCuWNGzcWGzZsKHe5sWPHCicnJ1FUVCSVDRs2TIwYMaLCffrtt9/KPGa8PC1atBBr166Vpp2cnMSoUaOk6bS0NAFAzJs3TypLSEgQAERaWpoQQojw8HBhZmYmzS8v/tLjm5ubKwwMDMTx48cV4nj77bfFG2+8IYQQYuDAgeKtt96qNG6iuo5tOIjqqOjoaBgbG6OoqAiFhYUYPHgw1q5dK813cnKClZWVNH3x4kU4OjrC0dFRKmvRogUaNGiAixcvon379uVuJzExEbm5ubCwsFAoz8vLw9WrVyuMr2XLltDW1pam7ezscP78eZX28eHDh1i0aBGio6Nx+/ZtFBUVIS8vr8wVDi8vL+n/pU8nLb1F83RZRkYGbG1tVYrhjz/+wOPHj9GzZ0+F8oKCArz66qsAgIkTJ2Lo0KE4c+YMevXqhYCAAHh7e6u0HaLajgkHUR3l6+uLdevWQVdXF/b29mUahRoZGSlMCyEgk8nKrKei8lIlJSWws7PD0aNHy8yrrB3Es/HIZDKUlJRUWL8877//Pvbv34+VK1fCzc0NcrkcQUFBKCgoqHBbpftSXpmq2396mZ9//hkODg4K8/T19QEAffv2xfXr1/Hzzz/j4MGD8Pf3x+TJk7Fy5UqVt0dUWzHhIKqjjIyM4ObmpnT9Fi1a4MaNG7h586Z0leOPP/5AVlYWmjdvDgDQ09NDcXGxwnJt2rRBeno6dHR04OzsrLb4lREXF4fg4GAMGTIEAJCbm1vpkzk1oUWLFtDX18eNGzfQrVu3CutZWVkhODgYwcHB8PHxwfvvv8+Eg+oVJhxEBADo0aMHvLy88Oabb+Kzzz5DUVERJk2ahG7duqFdu3YAnvT4uHbtGs6ePYtGjRrBxMQEPXr0QKdOnRAQEIDly5ejWbNmuH37Nvbs2YOAgABpWU1wc3NDVFQUBg4cCJlMhnnz5lXpKsWLMDExwaxZszBjxgyUlJSgS5cuyM7OxvHjx2FsbIyxY8di/vz5aNu2LVq2bIn8/HxER0dLSRxRfcFeKkQE4MlthZ07d6Jhw4bo2rUrevToAVdXV/zwww9SnaFDh6JPnz7w9fWFlZUVtmzZAplMhj179qBr164YN24cmjZtitdffx2pqalS2whNWbNmDRo2bAhvb28MHDgQvXv3Rps2bTS6zfJ8/PHHmD9/PsLCwtC8eXP07t0bu3fvhouLC4AnV4ZCQ0Ph5eWFrl27QltbG1u3bn3pcRJVJ5kQVewHRkRERKQkXuEgIiIijWPCQURERBrHhIOIiIg0jgkHERERaRwTDiIiItI4JhxERESkcUw4iIiISOOYcBAREZHGMeEgIiIijWPCQURERBrHhIOIiIg07v8BljUE6EkShBkAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 600x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Outlier cut-off for each protein family: {'Lysozyme_C': [137, 158], 'Lysozyme_G': [167, 217], 'Lysozyme_Ch': [216, 235]}\n",
      "Total number of protein sequences removed as outliers: 20\n",
      "Total number of protein sequences below cut-off 30: 0\n",
      "\n",
      "Contingency Matrix for KMeans:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>2</td>\n",
       "      <td>34</td>\n",
       "      <td>16</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>52</td>\n",
       "      <td>39</td>\n",
       "      <td>32</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Total\n",
       "Protein family                                        \n",
       "Lysozyme_C              2         34         16     52\n",
       "Lysozyme_Ch            50          0          4     54\n",
       "Lysozyme_G              0          5         12     17\n",
       "Total                  52         39         32    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for Agglomerative:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>53</td>\n",
       "      <td>54</td>\n",
       "      <td>16</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Total\n",
       "Protein family                                        \n",
       "Lysozyme_C             52          0          0     52\n",
       "Lysozyme_Ch             0         54          0     54\n",
       "Lysozyme_G              1          0         16     17\n",
       "Total                  53         54         16    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for Gaussian mixture:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>2</td>\n",
       "      <td>35</td>\n",
       "      <td>15</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>52</td>\n",
       "      <td>40</td>\n",
       "      <td>31</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Total\n",
       "Protein family                                        \n",
       "Lysozyme_C              2         35         15     52\n",
       "Lysozyme_Ch            50          0          4     54\n",
       "Lysozyme_G              0          5         12     17\n",
       "Total                  52         40         31    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for Spectral:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>54</td>\n",
       "      <td>51</td>\n",
       "      <td>18</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Total\n",
       "Protein family                                        \n",
       "Lysozyme_C              2         50          0     52\n",
       "Lysozyme_Ch            52          0          2     54\n",
       "Lysozyme_G              0          1         16     17\n",
       "Total                  54         51         18    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for Affinity propagation:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Cluster 3</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>17</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>17</td>\n",
       "      <td>34</td>\n",
       "      <td>19</td>\n",
       "      <td>53</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Cluster 3  Total\n",
       "Protein family                                                   \n",
       "Lysozyme_C             17         33          0          2     52\n",
       "Lysozyme_Ch             0          0          3         51     54\n",
       "Lysozyme_G              0          1         16          0     17\n",
       "Total                  17         34         19         53    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for DBSCAN:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Noise</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Noise  Total\n",
       "Protein family              \n",
       "Lysozyme_C         52     52\n",
       "Lysozyme_Ch        54     54\n",
       "Lysozyme_G         17     17\n",
       "Total             123    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for OPTICS:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Noise</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Cluster 3</th>\n",
       "      <th>Cluster 4</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>33</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>72</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Noise  Cluster 0  Cluster 1  Cluster 2  Cluster 3  Cluster 4  \\\n",
       "Protein family                                                                 \n",
       "Lysozyme_C         33          5          8          6          0          0   \n",
       "Lysozyme_Ch        38          0          0          0          0         16   \n",
       "Lysozyme_G          1          0          0          0         16          0   \n",
       "Total              72          5          8          6         16         16   \n",
       "\n",
       "Cluster         Total  \n",
       "Protein family         \n",
       "Lysozyme_C         52  \n",
       "Lysozyme_Ch        54  \n",
       "Lysozyme_G         17  \n",
       "Total             123  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for Mean-shift:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>121</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Cluster 1  Cluster 2  Total\n",
       "Protein family                                        \n",
       "Lysozyme_C             50          1          1     52\n",
       "Lysozyme_Ch            54          0          0     54\n",
       "Lysozyme_G             17          0          0     17\n",
       "Total                 121          1          1    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Contingency Matrix for BIRCH:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Cluster</th>\n",
       "      <th>Cluster 0</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lysozyme_C</th>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_Ch</th>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysozyme_G</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Cluster         Cluster 0  Total\n",
       "Protein family                  \n",
       "Lysozyme_C             52     52\n",
       "Lysozyme_Ch            54     54\n",
       "Lysozyme_G             17     17\n",
       "Total                 123    123"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of subclusters found 1 by BIRCH even though k=3\n",
      "\n",
      "Summary of clustering results for Lysozyme_CGCCh dataset:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>expected_clusters</th>\n",
       "      <th>predicted_clusters</th>\n",
       "      <th>total_features</th>\n",
       "      <th>total_samples</th>\n",
       "      <th>SC</th>\n",
       "      <th>CH</th>\n",
       "      <th>DB</th>\n",
       "      <th>FMI</th>\n",
       "      <th>F1</th>\n",
       "      <th>MCC</th>\n",
       "      <th>AMI</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clustering_method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KMeans</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.30</td>\n",
       "      <td>55.45</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Agglomerative</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.29</td>\n",
       "      <td>54.32</td>\n",
       "      <td>1.28</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gaussian mixture</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.30</td>\n",
       "      <td>55.39</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spectral</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.30</td>\n",
       "      <td>57.83</td>\n",
       "      <td>1.36</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Affinity propagation</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.30</td>\n",
       "      <td>48.01</td>\n",
       "      <td>1.53</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DBSCAN</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OPTICS</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>12.84</td>\n",
       "      <td>1.46</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mean-shift</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>0.07</td>\n",
       "      <td>1.67</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.55</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIRCH</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      expected_clusters  predicted_clusters  total_features  \\\n",
       "Clustering_method                                                             \n",
       "KMeans                                3                   3              50   \n",
       "Agglomerative                         3                   3              50   \n",
       "Gaussian mixture                      3                   3              50   \n",
       "Spectral                              3                   3              50   \n",
       "Affinity propagation                  3                   4              50   \n",
       "DBSCAN                                3                   1              50   \n",
       "OPTICS                                3                   6              50   \n",
       "Mean-shift                            3                   3              50   \n",
       "BIRCH                                 3                   1              50   \n",
       "\n",
       "                      total_samples    SC     CH    DB   FMI    F1   MCC   AMI  \n",
       "Clustering_method                                                               \n",
       "KMeans                          123  0.30  55.45  1.71  0.73  0.73  0.58  0.53  \n",
       "Agglomerative                   123  0.29  54.32  1.28  0.99  0.99  0.98  0.96  \n",
       "Gaussian mixture                123  0.30  55.39  1.71  0.74  0.73  0.59  0.53  \n",
       "Spectral                        123  0.30  57.83  1.36  0.93  0.93  0.89  0.83  \n",
       "Affinity propagation            123  0.30  48.01  1.53  0.81  0.80  0.72  0.72  \n",
       "DBSCAN                          123   NaN    NaN   NaN  0.62  0.56  0.00  0.00  \n",
       "OPTICS                          123 -0.00  12.84  1.46  0.53  0.53  0.24  0.46  \n",
       "Mean-shift                      123  0.07   1.67  0.82  0.61  0.55 -0.01 -0.00  \n",
       "BIRCH                           123   NaN    NaN   NaN  0.62  0.56  0.00  0.00  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Note: NaN values indicate that the number of clusters is 1, which is not suitable for calculating intrinsic evaluation metrics.\n",
      " -0.00 indicates a very small negative value that has been rounded to 0.00.\n"
     ]
    }
   ],
   "source": [
    "print(f'Clustering results for {dataset} dataset')\n",
    "# Preprocessing steps\n",
    "# 1) remove duplicates within protein family based on accession number\n",
    "# 2) remove duplicates between protein families based on accession number\n",
    "# 3) remove sequences with \"UZOBJX\" amino acid\n",
    "preprocessed_fasta = preprocessing_step(protein_families)\n",
    "\n",
    "# Optional preprocessing steps\n",
    "# 1) Remove outliers based on the lower and upper bounds for the protein sequences length using box-whisker plot\n",
    "# 2) Remove the sequence below assigned sequence length cut-off\n",
    "## a) This sequence length cut-off is done to get feature of same dimension size for clustering\n",
    "optional_preprocessing_fasta = optional_preprocessing_step(preprocessed_fasta, 30) # 30 is the cut-off for the protein sequence length\n",
    "\n",
    "# Feature extraction using pseudo amino acid composition (PAAC)\n",
    "# 1) lambda_ is the sequence order\n",
    "# 2) w is the weight factor\n",
    "PAAC = pseudo_amino_acid_composition(optional_preprocessing_fasta.len_filtered_fasta, 30, 0.05) # default lambda_ = 30 and w = 0.05\n",
    "\n",
    "# Clustering\n",
    "## 1) KMeans, 2) Agglomerative clustering, 3) Gaussian mixture, 4) Spectral clustering, 5) Affinity propagation, 6) DBSCAN, 7) OPTICS, 8) Mean-shift, 9) BIRCH\n",
    "# Evaluation metrics\n",
    "## 1) Intrinsic: Silhouette coefficient, Calinski-Harabasz index, Davies-Bouldin index\n",
    "## 2) Extrinsic: Fowlkes-Mallows index, F1 score, Matthews correlation coefficient, Adjusted mutual information  \n",
    "## 3) Contingency matrix\n",
    "clustering_results = clustering(PAAC.df, 3) # 3 is the number of clusters\n",
    "kmeans_results = clustering_results.kmeans()\n",
    "agglomerative_results = clustering_results.agglomerative()\n",
    "gaussian_mixture_results = clustering_results.gaussian_mixture()\n",
    "spectral_results = clustering_results.spectral()\n",
    "affinity_propagation_results = clustering_results.affinity_propagation()\n",
    "dbscan_results = clustering_results.dbscan()\n",
    "optics_results = clustering_results.optics()\n",
    "mean_shift_results = clustering_results.mean_shift()\n",
    "birch_results = clustering_results.birch()\n",
    "\n",
    "# Summary of clustering results\n",
    "clustering_summary = pd.DataFrame([kmeans_results, agglomerative_results, gaussian_mixture_results, spectral_results, \n",
    "                                   affinity_propagation_results, dbscan_results, optics_results, mean_shift_results, birch_results])\n",
    "# Set clustering_summary index to Clustering_method column and drop the original column\n",
    "clustering_summary.set_index('Clustering_method', inplace=True)\n",
    "print(f'\\nSummary of clustering results for {dataset} dataset:')\n",
    "display(clustering_summary)\n",
    "print('\\n')\n",
    "print('Note: NaN values indicate that the number of clusters is 1, which is not suitable for calculating intrinsic evaluation metrics.\\n -0.00 indicates a very small negative value that has been rounded to 0.00.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "simple",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
